<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 CAPÍTULO N° 5 - Clustering | Guía de Bioestadistica</title>
  <meta name="description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  <meta name="generator" content="bookdown 0.26 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 CAPÍTULO N° 5 - Clustering | Guía de Bioestadistica" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 CAPÍTULO N° 5 - Clustering | Guía de Bioestadistica" />
  
  <meta name="twitter:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  



<meta name="date" content="2022-08-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="capítulo-n-4---modelos-de-mezcla.html"/>
<link rel="next" href="footnotes-and-citations.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introdución</a></li>
<li class="chapter" data-level="2" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html"><i class="fa fa-check"></i><b>2</b> CAPÍTULO N° 2 - Modelamiento Estadístico</a>
<ul>
<li class="chapter" data-level="2.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#un-ejemplo-simple-de-modelado-estadístico"><i class="fa fa-check"></i><b>2.1</b> Un ejemplo simple de modelado estadístico</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#estimación-del-parámetro-de-la-distribución-de-poisson"><i class="fa fa-check"></i><b>2.1.1</b> Estimación del parámetro de la distribución de Poisson</a></li>
<li class="chapter" data-level="2.1.2" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#función-de-verosimilitud-de-λ"><i class="fa fa-check"></i><b>2.1.2</b> Función de verosimilitud de λ</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#distribuciones-binomiales-y-máxima-verosimilitud"><i class="fa fa-check"></i><b>2.2</b> Distribuciones binomiales y máxima verosimilitud</a></li>
<li class="chapter" data-level="2.3" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#más-casillasdatos-multinomiales"><i class="fa fa-check"></i><b>2.3</b> Más casillas:datos multinomiales</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#sesgo-de-nucleótidos"><i class="fa fa-check"></i><b>2.3.1</b> Sesgo de nucleótidos</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#la-distribución-chi-cuadrado"><i class="fa fa-check"></i><b>2.4</b> La distribución Chi-Cuadrado</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#intermezzo-cuantiles-y-gráfico-cuantil-cuantil"><i class="fa fa-check"></i><b>2.4.1</b> Intermezzo: cuantiles y gráfico cuantil-cuantil</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#regla-de-chargaff"><i class="fa fa-check"></i><b>2.5</b> Regla de Chargaff</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#dos-variables-categóricas"><i class="fa fa-check"></i><b>2.5.1</b> Dos variables categóricas</a></li>
<li class="chapter" data-level="2.5.2" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#daltonismo-y-sexo"><i class="fa fa-check"></i><b>2.5.2</b> <strong>Daltonismo y sexo</strong></a></li>
<li class="chapter" data-level="2.5.3" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#un-multinomio-especial-equilibrio-de-hardy-weinberg"><i class="fa fa-check"></i><b>2.5.3</b> Un multinomio especial: equilibrio de Hardy-Weinberg</a></li>
<li class="chapter" data-level="2.5.4" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#comparación-visual-con-el-equilibrio-de-hardy-weinberg"><i class="fa fa-check"></i><b>2.5.4</b> Comparación visual con el equilibrio de Hardy-Weinberg</a></li>
<li class="chapter" data-level="2.5.5" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#concatenación-de-varios-multinomios-motivos-de-secuencia-y-logos"><i class="fa fa-check"></i><b>2.5.5</b> Concatenación de varios multinomios: motivos de secuencia y logos</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#modelado-de-dependencias-secuenciales-cadenas-de-markov"><i class="fa fa-check"></i><b>2.6</b> Modelado de dependencias secuenciales: cadenas de Markov</a></li>
<li class="chapter" data-level="2.7" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#pensamiento-bayesiano"><i class="fa fa-check"></i><b>2.7</b> Pensamiento bayesiano</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#haplotipos"><i class="fa fa-check"></i><b>2.7.1</b> Haplotipos</a></li>
<li class="chapter" data-level="2.7.2" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#estudio-de-simulación-del-paradigma-bayesiano-para-el-binomio"><i class="fa fa-check"></i><b>2.7.2</b> Estudio de simulación del paradigma bayesiano para el binomio</a></li>
<li class="chapter" data-level="2.7.3" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#la-distribución-de-y"><i class="fa fa-check"></i><b>2.7.3</b> La distribución de Y</a></li>
<li class="chapter" data-level="2.7.4" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#la-distribución-posterior-también-es-una-beta."><i class="fa fa-check"></i><b>2.7.4</b> La distribución posterior también es una beta.</a></li>
<li class="chapter" data-level="2.7.5" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#declaraciones-de-confianza-para-el-parámetro-de-proporción"><i class="fa fa-check"></i><b>2.7.5</b> Declaraciones de confianza para el parámetro de proporción</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#ejemplo-aparición-de-un-patrón-de-nucleótidos-en-un-genoma"><i class="fa fa-check"></i><b>2.8</b> Ejemplo: aparición de un patrón de nucleótidos en un genoma</a>
<ul>
<li class="chapter" data-level="2.8.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#modelado-en-el-caso-de-dependencias"><i class="fa fa-check"></i><b>2.8.1</b> Modelado en el caso de dependencias</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#resumen-de-este-capítulo"><i class="fa fa-check"></i><b>2.9</b> Resumen de este capítulo</a></li>
<li class="chapter" data-level="2.10" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#ejercicios"><i class="fa fa-check"></i><b>2.10</b> Ejercicios</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#pregunta-n-1"><i class="fa fa-check"></i><b>2.10.1</b> Pregunta N° 1</a></li>
<li class="chapter" data-level="2.10.2" data-path="capítulo-n-2---modelamiento-estadístico.html"><a href="capítulo-n-2---modelamiento-estadístico.html#pregunta-n-2"><i class="fa fa-check"></i><b>2.10.2</b> pregunta N° 2</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><i class="fa fa-check"></i><b>3</b> CAPÍTULO N° 3 - Gráficos de alta calidad en R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#objetivos-del-capítulo"><i class="fa fa-check"></i><b>3.1</b> Objetivos del capítulo</a></li>
<li class="chapter" data-level="3.2" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#trazado-base-r"><i class="fa fa-check"></i><b>3.2</b> Trazado base R</a></li>
<li class="chapter" data-level="3.3" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#un-conjunto-de-datos-de-ejemplo"><i class="fa fa-check"></i><b>3.3</b> Un conjunto de datos de ejemplo</a></li>
<li class="chapter" data-level="3.4" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#ggplot2"><i class="fa fa-check"></i><b>3.4</b> ggplot2</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#estética"><i class="fa fa-check"></i><b>3.4.1</b> Estética</a></li>
<li class="chapter" data-level="3.4.2" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#flujo-de-datos"><i class="fa fa-check"></i><b>3.4.2</b> Flujo de datos</a></li>
<li class="chapter" data-level="3.4.3" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#guardar-cifras"><i class="fa fa-check"></i><b>3.4.3</b> Guardar cifras</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#la-gramática-de-los-gráficos"><i class="fa fa-check"></i><b>3.5</b> La gramática de los gráficos</a></li>
<li class="chapter" data-level="3.6" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#visualización-de-datos-en-1d"><i class="fa fa-check"></i><b>3.6</b> Visualización de datos en 1D</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#diagramas-de-barras"><i class="fa fa-check"></i><b>3.6.1</b> Diagramas de barras</a></li>
<li class="chapter" data-level="3.6.2" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#diagramas-de-caja"><i class="fa fa-check"></i><b>3.6.2</b> Diagramas de caja</a></li>
<li class="chapter" data-level="3.6.3" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#violín"><i class="fa fa-check"></i><b>3.6.3</b> violín</a></li>
<li class="chapter" data-level="3.6.4" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#diagramas-de-densidad"><i class="fa fa-check"></i><b>3.6.4</b> Diagramas de densidad</a></li>
<li class="chapter" data-level="3.6.5" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#gráficos-ecdf"><i class="fa fa-check"></i><b>3.6.5</b> Gráficos ECDF</a></li>
<li class="chapter" data-level="3.6.6" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#el-efecto-de-las-transformaciones-sobre-las-densidades"><i class="fa fa-check"></i><b>3.6.6</b> El efecto de las transformaciones sobre las densidades</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#visualización-de-datos-en-2d-diagramas-de-dispersión"><i class="fa fa-check"></i><b>3.7</b> Visualización de datos en 2D: diagramas de dispersión</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#representada-como-un-gráfico-de-contorno-de-la-estimación-de-densidad-2d."><i class="fa fa-check"></i><b>3.7.1</b> Representada como un gráfico de contorno de la estimación de densidad 2D.</a></li>
<li class="chapter" data-level="3.7.2" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#trazar-formas"><i class="fa fa-check"></i><b>3.7.2</b> Trazar formas</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#visualización-de-más-de-dos-dimensiones"><i class="fa fa-check"></i><b>3.8</b> Visualización de más de dos dimensiones</a>
<ul>
<li class="chapter" data-level="3.8.1" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#facetado"><i class="fa fa-check"></i><b>3.8.1</b> Facetado</a></li>
<li class="chapter" data-level="3.8.2" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#gráficos-interactivos"><i class="fa fa-check"></i><b>3.8.2</b> Gráficos interactivos</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#color"><i class="fa fa-check"></i><b>3.9</b> Color</a></li>
<li class="chapter" data-level="3.10" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#mapas-de-calor"><i class="fa fa-check"></i><b>3.10</b> Mapas de calor</a></li>
<li class="chapter" data-level="3.11" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#transformaciones-de-datos"><i class="fa fa-check"></i><b>3.11</b> Transformaciones de datos</a></li>
<li class="chapter" data-level="3.12" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#símbolos-matemáticos-y-otras-fuentes"><i class="fa fa-check"></i><b>3.12</b> Símbolos matemáticos y otras fuentes</a></li>
<li class="chapter" data-level="3.13" data-path="capítulo-n-3---gráficos-de-alta-calidad-en-r.html"><a href="capítulo-n-3---gráficos-de-alta-calidad-en-r.html#datos-genómicos"><i class="fa fa-check"></i><b>3.13</b> Datos genómicos</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html"><i class="fa fa-check"></i><b>4</b> CAPÍTULO N° 4 - Modelos de mezcla</a>
<ul>
<li class="chapter" data-level="4.1" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#objetivos-de-este-capítulo"><i class="fa fa-check"></i><b>4.1</b> Objetivos de este capítulo</a></li>
<li class="chapter" data-level="4.2" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#mezclas-finitas"><i class="fa fa-check"></i><b>4.2</b> Mezclas finitas</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#ejemplos-sencillos-y-experimentos-informáticos"><i class="fa fa-check"></i><b>4.2.1</b> Ejemplos sencillos y experimentos informáticos</a></li>
<li class="chapter" data-level="4.2.2" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#descubriendo-las-etiquetas-de-clase-ocultas"><i class="fa fa-check"></i><b>4.2.2</b> Descubriendo las etiquetas de clase ocultas</a></li>
<li class="chapter" data-level="4.2.3" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#modelos-para-datos-inflados-a-cero"><i class="fa fa-check"></i><b>4.2.3</b> Modelos para datos inflados a cero</a></li>
<li class="chapter" data-level="4.2.4" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#más-de-dos-componentes"><i class="fa fa-check"></i><b>4.2.4</b> Más de dos componentes</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#distribuciones-empíricas-y-bootstrap-no-paramétrico"><i class="fa fa-check"></i><b>4.3</b> Distribuciones empíricas y bootstrap no paramétrico</a></li>
<li class="chapter" data-level="4.4" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#mezclas-infinitas"><i class="fa fa-check"></i><b>4.4</b> Mezclas infinitas</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#mezcla-infinita-de-normales"><i class="fa fa-check"></i><b>4.4.1</b> Mezcla infinita de normales</a></li>
<li class="chapter" data-level="4.4.2" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#mezclas-infinitas-de-variables-de-poisson"><i class="fa fa-check"></i><b>4.4.2</b> Mezclas infinitas de variables de Poisson</a></li>
<li class="chapter" data-level="4.4.3" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#transformaciones-estabilizadoras-de-varianza"><i class="fa fa-check"></i><b>4.4.3</b> Transformaciones estabilizadoras de varianza</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#resumen-de-este-capítulo-1"><i class="fa fa-check"></i><b>4.5</b> Resumen de este capítulo</a></li>
<li class="chapter" data-level="4.6" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#ejercicios-1"><i class="fa fa-check"></i><b>4.6</b> Ejercicios</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#ejercicio-4.1"><i class="fa fa-check"></i><b>4.6.1</b> Ejercicio 4.1</a></li>
<li class="chapter" data-level="4.6.2" data-path="capítulo-n-4---modelos-de-mezcla.html"><a href="capítulo-n-4---modelos-de-mezcla.html#ejercicio-4.2"><i class="fa fa-check"></i><b>4.6.2</b> Ejercicio 4.2</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html"><i class="fa fa-check"></i><b>5</b> CAPÍTULO N° 5 - Clustering</a>
<ul>
<li class="chapter" data-level="5.1" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#objetivos-de-este-capítulo-1"><i class="fa fa-check"></i><b>5.1</b> Objetivos de este capítulo</a></li>
<li class="chapter" data-level="5.2" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#qué-son-los-datos-y-por-qué-los-agrupamos"><i class="fa fa-check"></i><b>5.2</b> ¿Qué son los datos y por qué los agrupamos?</a></li>
<li class="chapter" data-level="5.3" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#cálculos-relacionados-con-distancias-en-r"><i class="fa fa-check"></i><b>5.3</b> 5.3.1 Cálculos relacionados con distancias en R</a></li>
<li class="chapter" data-level="5.4" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#detección-de-mezcla-no-paramétrica"><i class="fa fa-check"></i><b>5.4</b> Detección de mezcla no paramétrica</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#k-métodos-k--significa-k--medoides-y-pam"><i class="fa fa-check"></i><b>5.4.1</b> K-métodos: k -significa, k -medoides y PAM</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#ejemplos-de-agrupamiento-citometría-de-flujo-y-citometría-de-masas"><i class="fa fa-check"></i><b>5.5</b> Ejemplos de agrupamiento: citometría de flujo y citometría de masas</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#citometría-de-flujo-y-citometría-de-masas"><i class="fa fa-check"></i><b>5.5.1</b> Citometría de flujo y citometría de masas</a></li>
<li class="chapter" data-level="5.5.2" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#preprocesamiento-de-datos"><i class="fa fa-check"></i><b>5.5.2</b> Preprocesamiento de datos</a></li>
<li class="chapter" data-level="5.5.3" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#agrupamiento-basado-en-la-densidad"><i class="fa fa-check"></i><b>5.5.3</b> Agrupamiento basado en la densidad</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#agrupamiento-jerárquico"><i class="fa fa-check"></i><b>5.6</b> Agrupamiento jerárquico</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#cómo-calcular-des-similitudes-entre-clústeres-agregados"><i class="fa fa-check"></i><b>5.6.1</b> ¿Cómo calcular (des) similitudes entre clústeres agregados?</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#validación-y-elección-del-número-de-clústeres"><i class="fa fa-check"></i><b>5.7</b> Validación y elección del número de clústeres</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#uso-de-la-estadística-de-la-brecha"><i class="fa fa-check"></i><b>5.7.1</b> Uso de la estadística de la brecha</a></li>
<li class="chapter" data-level="5.7.2" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#validación-de-clúster-mediante-bootstrap"><i class="fa fa-check"></i><b>5.7.2</b> Validación de clúster mediante bootstrap</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#el-agrupamiento-como-medio-para-eliminar-el-ruido"><i class="fa fa-check"></i><b>5.8</b> El agrupamiento como medio para eliminar el ruido</a>
<ul>
<li class="chapter" data-level="5.8.1" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#observaciones-ruidosas-con-diferentes-frecuencias-de-línea-de-base"><i class="fa fa-check"></i><b>5.8.1</b> Observaciones ruidosas con diferentes frecuencias de línea de base</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="capítulo-n-5---clustering.html"><a href="capítulo-n-5---clustering.html#ejercicios-2"><i class="fa fa-check"></i><b>5.9</b> Ejercicios</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="footnotes-and-citations.html"><a href="footnotes-and-citations.html"><i class="fa fa-check"></i><b>6</b> Footnotes and citations</a>
<ul>
<li class="chapter" data-level="6.1" data-path="footnotes-and-citations.html"><a href="footnotes-and-citations.html#footnotes"><i class="fa fa-check"></i><b>6.1</b> Footnotes</a></li>
<li class="chapter" data-level="6.2" data-path="footnotes-and-citations.html"><a href="footnotes-and-citations.html#citations"><i class="fa fa-check"></i><b>6.2</b> Citations</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="blocks.html"><a href="blocks.html"><i class="fa fa-check"></i><b>7</b> Blocks</a>
<ul>
<li class="chapter" data-level="7.1" data-path="blocks.html"><a href="blocks.html#equations"><i class="fa fa-check"></i><b>7.1</b> Equations</a></li>
<li class="chapter" data-level="7.2" data-path="blocks.html"><a href="blocks.html#theorems-and-proofs"><i class="fa fa-check"></i><b>7.2</b> Theorems and proofs</a></li>
<li class="chapter" data-level="7.3" data-path="blocks.html"><a href="blocks.html#callout-blocks"><i class="fa fa-check"></i><b>7.3</b> Callout blocks</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="objetivos-de-este-capítulo-2.html"><a href="objetivos-de-este-capítulo-2.html"><i class="fa fa-check"></i><b>8</b> 6.1 Objetivos de este Capítulo</a>
<ul>
<li class="chapter" data-level="8.1" data-path="objetivos-de-este-capítulo-2.html"><a href="objetivos-de-este-capítulo-2.html#corrección-de-bonferroni"><i class="fa fa-check"></i><b>8.1</b> 6.8.1 Corrección de Bonferroni</a></li>
<li class="chapter" data-level="8.2" data-path="objetivos-de-este-capítulo-2.html"><a href="objetivos-de-este-capítulo-2.html#el-histograma-del-valor-p"><i class="fa fa-check"></i><b>8.2</b> 6.9.1 El histograma del valor p</a></li>
<li class="chapter" data-level="8.3" data-path="objetivos-de-este-capítulo-2.html"><a href="objetivos-de-este-capítulo-2.html#el-algoritmo-de-benjamini-hochberg-para-controlar-el-fdr"><i class="fa fa-check"></i><b>8.3</b> 6.9.2 El algoritmo de Benjamini-Hochberg para controlar el FDR</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="objetivos-de-este-capítulo-2.html"><a href="objetivos-de-este-capítulo-2.html#pregunta"><i class="fa fa-check"></i><b>8.3.1</b> Pregunta</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html"><i class="fa fa-check"></i><b>9</b> 7.1 Objetivos de este capítulo</a>
<ul>
<li class="chapter" data-level="9.1" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#resúmenes-y-preparación-de-datos-de-baja-dimensión"><i class="fa fa-check"></i><b>9.1</b> 7.2.1 Resúmenes y preparación de datos de baja dimensión</a></li>
<li class="chapter" data-level="9.2" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#proyecciones-de-menor-dimensión"><i class="fa fa-check"></i><b>9.2</b> 7.3.1 Proyecciones de menor dimensión</a></li>
<li class="chapter" data-level="9.3" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#cómo-resumimos-los-datos-bidimensionales-en-una-línea"><i class="fa fa-check"></i><b>9.3</b> 7.3.2 ¿Cómo resumimos los datos bidimensionales en una línea?</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#regresión-de-una-variable-sobre-la-otra"><i class="fa fa-check"></i><b>9.3.1</b> Regresión de una variable sobre la otra</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#matrices-de-rango-uno"><i class="fa fa-check"></i><b>9.4</b> 7.6.1 Matrices de rango uno</a></li>
<li class="chapter" data-level="9.5" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#cómo-encontramos-tal-descomposición-de-una-manera-única"><i class="fa fa-check"></i><b>9.5</b> 7.6.2 ¿Cómo encontramos tal descomposición de una manera única?</a></li>
<li class="chapter" data-level="9.6" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#pca-de-los-datos-de-las-tortugas"><i class="fa fa-check"></i><b>9.6</b> 7.7.1 PCA de los datos de las tortugas</a></li>
<li class="chapter" data-level="9.7" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#un-análisis-completo-los-atletas-de-decatlón"><i class="fa fa-check"></i><b>9.7</b> 7.7.2 Un análisis completo: los atletas de decatlón</a></li>
<li class="chapter" data-level="9.8" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#cómo-elegir-k-el-número-de-dimensiones"><i class="fa fa-check"></i><b>9.8</b> 7.7.3 Cómo elegir k , el número de dimensiones ?</a></li>
<li class="chapter" data-level="9.9" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#análisis-de-datos-de-espectroscopia-de-masas"><i class="fa fa-check"></i><b>9.9</b> 7.8.1 Análisis de datos de espectroscopia de masas</a></li>
<li class="chapter" data-level="9.10" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#biplots-y-escalado"><i class="fa fa-check"></i><b>9.10</b> 7.8.2 Biplots y escalado</a></li>
<li class="chapter" data-level="9.11" data-path="objetivos-de-este-capítulo-3.html"><a href="objetivos-de-este-capítulo-3.html#un-ejemplo-de-pca-ponderado"><i class="fa fa-check"></i><b>9.11</b> 7.8.3 Un ejemplo de PCA ponderado</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="sharing-your-book.html"><a href="sharing-your-book.html"><i class="fa fa-check"></i><b>10</b> Sharing your book</a>
<ul>
<li class="chapter" data-level="10.1" data-path="sharing-your-book.html"><a href="sharing-your-book.html#publishing"><i class="fa fa-check"></i><b>10.1</b> Publishing</a></li>
<li class="chapter" data-level="10.2" data-path="sharing-your-book.html"><a href="sharing-your-book.html#pages"><i class="fa fa-check"></i><b>10.2</b> 404 pages</a></li>
<li class="chapter" data-level="10.3" data-path="sharing-your-book.html"><a href="sharing-your-book.html#metadata-for-sharing"><i class="fa fa-check"></i><b>10.3</b> Metadata for sharing</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Guía de Bioestadistica</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="capítulo-n-5---clustering" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> CAPÍTULO N° 5 - Clustering<a href="capítulo-n-5---clustering.html#capítulo-n-5---clustering" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="objetivos-de-este-capítulo-1" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Objetivos de este capítulo<a href="capítulo-n-5---clustering.html#objetivos-de-este-capítulo-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p>Estudie los diferentes tipos de datos que se pueden agrupar de manera beneficiosa.</p></li>
<li><p>Ver medidas de (des)similitud y distancias que nos ayudan a definir clusters.</p></li>
<li><p>Descubra el agrupamiento oculto o latente dividiendo los datos en más reducidos.</p></li>
<li><p>Utilice la agrupación cuando se le den biomarcadores en cada una de cientos de miles de células. Veremos que, por ejemplo, las células inmunitarias se pueden agrupar naturalmente en subpoblaciones estrechas.</p></li>
<li><p>Ejecutar algoritmos no paramétricos como <span class="math inline">\(k\)</span>-significa , <span class="math inline">\(k\)</span>-medoides en datos reales de una sola celda.</p></li>
<li><p>Experimente con enfoques recursivos de agrupamiento que combinen observaciones y grupos en una jerarquía de conjuntos; estos métodos se conocen como agrupamiento jerárquico .</p></li>
<li><p>Estudie cómo validar clústeres a través de enfoques de arranque basados ​​en remuestreo, que demostraremos en un conjunto de datos de una sola celda.</p></li>
</ul>
</div>
<div id="qué-son-los-datos-y-por-qué-los-agrupamos" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> ¿Qué son los datos y por qué los agrupamos?<a href="capítulo-n-5---clustering.html#qué-son-los-datos-y-por-qué-los-agrupamos" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="capítulo-n-5---clustering.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PAQUETES </span></span>
<span id="cb1-2"><a href="capítulo-n-5---clustering.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pacman)</span>
<span id="cb1-3"><a href="capítulo-n-5---clustering.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;MASS&quot;</span>,<span class="st">&quot;RColorBrewer&quot;</span>,<span class="st">&quot;ggplot2&quot;</span>)</span></code></pre></div>
<p><strong>Distancia basada en correlación</strong></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="capítulo-n-5---clustering.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">101</span>)</span>
<span id="cb2-2"><a href="capítulo-n-5---clustering.html#cb2-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">60000</span></span>
<span id="cb2-3"><a href="capítulo-n-5---clustering.html#cb2-3" aria-hidden="true" tabindex="-1"></a>S1<span class="ot">=</span><span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,.<span class="dv">72</span>,.<span class="dv">72</span>,<span class="dv">1</span>), <span class="at">ncol=</span><span class="dv">2</span>)</span>
<span id="cb2-4"><a href="capítulo-n-5---clustering.html#cb2-4" aria-hidden="true" tabindex="-1"></a>S2<span class="ot">=</span><span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">1.5</span>,<span class="sc">-</span><span class="fl">0.6</span>,<span class="sc">-</span><span class="fl">0.6</span>,<span class="fl">1.5</span>),<span class="at">ncol=</span><span class="dv">2</span>)</span>
<span id="cb2-5"><a href="capítulo-n-5---clustering.html#cb2-5" aria-hidden="true" tabindex="-1"></a>mu1<span class="ot">=</span><span class="fu">c</span>(.<span class="dv">5</span>,<span class="fl">2.5</span>)</span>
<span id="cb2-6"><a href="capítulo-n-5---clustering.html#cb2-6" aria-hidden="true" tabindex="-1"></a>mu2<span class="ot">=</span><span class="fu">c</span>(<span class="fl">6.5</span>,<span class="dv">4</span>)</span>
<span id="cb2-7"><a href="capítulo-n-5---clustering.html#cb2-7" aria-hidden="true" tabindex="-1"></a>X1 <span class="ot">=</span> <span class="fu">mvrnorm</span>(n, <span class="at">mu=</span><span class="fu">c</span>(.<span class="dv">5</span>,<span class="fl">2.5</span>), <span class="at">Sigma=</span><span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,.<span class="dv">72</span>,.<span class="dv">72</span>,<span class="dv">1</span>), <span class="at">ncol=</span><span class="dv">2</span>))</span>
<span id="cb2-8"><a href="capítulo-n-5---clustering.html#cb2-8" aria-hidden="true" tabindex="-1"></a>X2 <span class="ot">=</span> <span class="fu">mvrnorm</span>(n,<span class="at">mu=</span><span class="fu">c</span>(<span class="fl">6.5</span>,<span class="dv">4</span>), <span class="at">Sigma=</span><span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">1.5</span>,<span class="sc">-</span><span class="fl">0.6</span>,<span class="sc">-</span><span class="fl">0.6</span>,<span class="fl">1.5</span>),<span class="at">ncol=</span><span class="dv">2</span>))</span>
<span id="cb2-9"><a href="capítulo-n-5---clustering.html#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A color palette from blue to yellow to red</span></span>
<span id="cb2-10"><a href="capítulo-n-5---clustering.html#cb2-10" aria-hidden="true" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">11</span></span>
<span id="cb2-11"><a href="capítulo-n-5---clustering.html#cb2-11" aria-hidden="true" tabindex="-1"></a>my.cols <span class="ot">&lt;-</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(k, <span class="st">&quot;RdYlBu&quot;</span>))</span>
<span id="cb2-12"><a href="capítulo-n-5---clustering.html#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(X1, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">4</span>,<span class="dv">12</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>,<span class="dv">9</span>), <span class="at">xlab=</span><span class="st">&quot;Orange&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Red&quot;</span>, <span class="at">pch=</span><span class="st">&#39;.&#39;</span>, <span class="at">cex=</span><span class="dv">1</span>)</span>
<span id="cb2-13"><a href="capítulo-n-5---clustering.html#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(X2, <span class="at">pch=</span><span class="st">&#39;.&#39;</span>, <span class="at">cex=</span><span class="dv">1</span>)</span>
<span id="cb2-14"><a href="capítulo-n-5---clustering.html#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Draw the colored contour lines</span></span>
<span id="cb2-15"><a href="capítulo-n-5---clustering.html#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># compute 2D kernel density, see MASS book, pp. 130-131</span></span>
<span id="cb2-16"><a href="capítulo-n-5---clustering.html#cb2-16" aria-hidden="true" tabindex="-1"></a>z1 <span class="ot">=</span> <span class="fu">kde2d</span>(X1[,<span class="dv">1</span>], X1[,<span class="dv">2</span>], <span class="at">n=</span><span class="dv">50</span>)</span>
<span id="cb2-17"><a href="capítulo-n-5---clustering.html#cb2-17" aria-hidden="true" tabindex="-1"></a>z2 <span class="ot">=</span> <span class="fu">kde2d</span>(X2[,<span class="dv">1</span>], X2[,<span class="dv">2</span>], <span class="at">n=</span><span class="dv">50</span>)</span>
<span id="cb2-18"><a href="capítulo-n-5---clustering.html#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="fu">contour</span>(z1, <span class="at">drawlabels=</span><span class="cn">FALSE</span>, <span class="at">nlevels=</span>k, <span class="at">col=</span>my.cols, <span class="at">add=</span><span class="cn">TRUE</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb2-19"><a href="capítulo-n-5---clustering.html#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="fu">contour</span>(z2, <span class="at">drawlabels=</span><span class="cn">FALSE</span>, <span class="at">nlevels=</span>k, <span class="at">col=</span>my.cols, <span class="at">add=</span><span class="cn">TRUE</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb2-20"><a href="capítulo-n-5---clustering.html#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="fl">3.2</span>,<span class="dv">2</span>,<span class="at">pch=</span><span class="dv">20</span>,<span class="at">cex=</span><span class="fl">2.2</span>,<span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb2-21"><a href="capítulo-n-5---clustering.html#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="fl">3.2</span>,<span class="fl">6.5</span>),<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>),<span class="at">col=</span><span class="st">&quot;red&quot;</span>,<span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb2-22"><a href="capítulo-n-5---clustering.html#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="fl">3.2</span>,.<span class="dv">5</span>),<span class="fu">c</span>(<span class="dv">2</span>,<span class="fl">2.5</span>),<span class="at">col=</span><span class="st">&quot;red&quot;</span>,<span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
<div id="cálculos-relacionados-con-distancias-en-r" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> 5.3.1 Cálculos relacionados con distancias en R<a href="capítulo-n-5---clustering.html#cálculos-relacionados-con-distancias-en-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="capítulo-n-5---clustering.html#cb3-1" aria-hidden="true" tabindex="-1"></a>mx  <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb3-2"><a href="capítulo-n-5---clustering.html#cb3-2" aria-hidden="true" tabindex="-1"></a>my  <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb3-3"><a href="capítulo-n-5---clustering.html#cb3-3" aria-hidden="true" tabindex="-1"></a>mz  <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb3-4"><a href="capítulo-n-5---clustering.html#cb3-4" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">=</span> <span class="fu">rbind</span>(mx, my, mz)</span>
<span id="cb3-5"><a href="capítulo-n-5---clustering.html#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dist</span>(mat)</span></code></pre></div>
<pre><code>##          mx       my
## my 1.732051         
## mz 2.000000 1.732051</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="capítulo-n-5---clustering.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dist</span>(mat, <span class="at">method =</span> <span class="st">&quot;binary&quot;</span>)</span></code></pre></div>
<pre><code>##           mx        my
## my 0.6000000          
## mz 0.6666667 0.5000000</code></pre>
<p>Puede acceder al conjunto de datos <a href="">aquí</a></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="capítulo-n-5---clustering.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;../data/Morder.RData&quot;</span>)</span>
<span id="cb7-2"><a href="capítulo-n-5---clustering.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="fu">sum</span>((Morder[<span class="dv">1</span>, ] <span class="sc">-</span> Morder[<span class="dv">2</span>, ])<span class="sc">^</span><span class="dv">2</span>))</span></code></pre></div>
<pre><code>## [1] 5.593667</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="capítulo-n-5---clustering.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.matrix</span>(<span class="fu">dist</span>(Morder))[<span class="dv">2</span>, <span class="dv">1</span>]</span></code></pre></div>
<pre><code>## [1] 5.593667</code></pre>
<p>Veamos cómo calcularíamos la distancia de Jaccard que definimos anteriormente entre las cepas de VIH.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="capítulo-n-5---clustering.html#cb11-1" aria-hidden="true" tabindex="-1"></a>mut <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">&quot;../data/HIVmutations.csv&quot;</span>)</span>
<span id="cb11-2"><a href="capítulo-n-5---clustering.html#cb11-2" aria-hidden="true" tabindex="-1"></a>mut[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">16</span>]</span></code></pre></div>
<pre><code>##   p32I p33F p34Q p35G p43T p46I p46L
## 1    0    1    0    0    0    0    0
## 2    0    1    0    0    0    1    0
## 3    0    1    0    0    0    0    0</code></pre>
</div>
<div id="detección-de-mezcla-no-paramétrica" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> Detección de mezcla no paramétrica<a href="capítulo-n-5---clustering.html#detección-de-mezcla-no-paramétrica" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="k-métodos-k--significa-k--medoides-y-pam" class="section level3 hasAnchor" number="5.4.1">
<h3><span class="header-section-number">5.4.1</span> K-métodos: k -significa, k -medoides y PAM<a href="capítulo-n-5---clustering.html#k-métodos-k--significa-k--medoides-y-pam" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="capítulo-n-5---clustering.html#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;vegan&quot;</span>)</span>
<span id="cb13-2"><a href="capítulo-n-5---clustering.html#cb13-2" aria-hidden="true" tabindex="-1"></a>mutJ <span class="ot">=</span> <span class="fu">vegdist</span>(mut, <span class="st">&quot;jaccard&quot;</span>)</span>
<span id="cb13-3"><a href="capítulo-n-5---clustering.html#cb13-3" aria-hidden="true" tabindex="-1"></a>mutC <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">2</span> <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> <span class="fu">cor</span>(<span class="fu">t</span>(mut))))</span>
<span id="cb13-4"><a href="capítulo-n-5---clustering.html#cb13-4" aria-hidden="true" tabindex="-1"></a>mutJ</span></code></pre></div>
<pre><code>##           1         2         3         4
## 2 0.8000000                              
## 3 0.7500000 0.8888889                    
## 4 0.9000000 0.7777778 0.8461538          
## 5 1.0000000 0.8000000 0.8888889 0.9000000</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="capítulo-n-5---clustering.html#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.dist</span>(mutC)</span></code></pre></div>
<pre><code>##          1        2        3        4
## 2 1.186342                           
## 3 1.104026 1.302931                  
## 4 1.318368 1.133893 1.298780         
## 5 1.452966 1.186342 1.302931 1.318368</code></pre>
<p>Ejemplo de ejecución de la k -significa algoritmo. Los centros (círculos negros) y grupos (colores) iniciales elegidos al azar se muestran en el panel superior. Los miembros del grupo se asignan en función de su distancia a los centros. En cada iteración, los centros de grupo se redefinen y los puntos se reasignan a los centros de clúster.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="capítulo-n-5---clustering.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">248811</span>)</span>
<span id="cb17-2"><a href="capítulo-n-5---clustering.html#cb17-2" aria-hidden="true" tabindex="-1"></a>Xmat <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">runif</span>(<span class="dv">100</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb17-3"><a href="capítulo-n-5---clustering.html#cb17-3" aria-hidden="true" tabindex="-1"></a>nk <span class="ot">=</span> <span class="dv">3</span></span>
<span id="cb17-4"><a href="capítulo-n-5---clustering.html#cb17-4" aria-hidden="true" tabindex="-1"></a>cents <span class="ot">=</span> Xmat[<span class="fu">sample</span>(<span class="fu">nrow</span>(Xmat), nk, <span class="at">replace =</span> <span class="cn">FALSE</span>), ]</span>
<span id="cb17-5"><a href="capítulo-n-5---clustering.html#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># default distance: Euclidean</span></span>
<span id="cb17-6"><a href="capítulo-n-5---clustering.html#cb17-6" aria-hidden="true" tabindex="-1"></a>dist1 <span class="ot">=</span> <span class="cf">function</span>(vec){<span class="fu">dist</span>(<span class="fu">rbind</span>(vec, cents[<span class="dv">1</span>,]))}</span>
<span id="cb17-7"><a href="capítulo-n-5---clustering.html#cb17-7" aria-hidden="true" tabindex="-1"></a>dist2 <span class="ot">=</span> <span class="cf">function</span>(vec){<span class="fu">dist</span>(<span class="fu">rbind</span>(vec, cents[<span class="dv">2</span>,]))}</span>
<span id="cb17-8"><a href="capítulo-n-5---clustering.html#cb17-8" aria-hidden="true" tabindex="-1"></a>dist3 <span class="ot">=</span> <span class="cf">function</span>(vec){<span class="fu">dist</span>(<span class="fu">rbind</span>(vec, cents[<span class="dv">3</span>,]))}</span>
<span id="cb17-9"><a href="capítulo-n-5---clustering.html#cb17-9" aria-hidden="true" tabindex="-1"></a>dists123 <span class="ot">=</span> <span class="fu">cbind</span>(<span class="fu">apply</span>(Xmat, <span class="dv">1</span>, dist1),</span>
<span id="cb17-10"><a href="capítulo-n-5---clustering.html#cb17-10" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">apply</span>(Xmat, <span class="dv">1</span>, dist2),</span>
<span id="cb17-11"><a href="capítulo-n-5---clustering.html#cb17-11" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">apply</span>(Xmat, <span class="dv">1</span>, dist3))</span>
<span id="cb17-12"><a href="capítulo-n-5---clustering.html#cb17-12" aria-hidden="true" tabindex="-1"></a>clust0 <span class="ot">=</span> <span class="fu">apply</span>(dists123, <span class="dv">1</span>, which.min)</span>
<span id="cb17-13"><a href="capítulo-n-5---clustering.html#cb17-13" aria-hidden="true" tabindex="-1"></a>out1 <span class="ot">=</span> <span class="fu">kmeans</span>(Xmat, cents, <span class="at">iter.max=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## Warning: did not converge in 1 iteration</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="capítulo-n-5---clustering.html#cb19-1" aria-hidden="true" tabindex="-1"></a>out2 <span class="ot">=</span> <span class="fu">kmeans</span>(Xmat, cents, <span class="at">iter.max=</span><span class="dv">3</span>)</span>
<span id="cb19-2"><a href="capítulo-n-5---clustering.html#cb19-2" aria-hidden="true" tabindex="-1"></a>data0 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">x =</span> Xmat[,<span class="dv">1</span>],</span>
<span id="cb19-3"><a href="capítulo-n-5---clustering.html#cb19-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y =</span> Xmat[,<span class="dv">2</span>],</span>
<span id="cb19-4"><a href="capítulo-n-5---clustering.html#cb19-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster =</span> <span class="fu">as.factor</span>(clust0))</span>
<span id="cb19-5"><a href="capítulo-n-5---clustering.html#cb19-5" aria-hidden="true" tabindex="-1"></a>data1 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">x =</span> Xmat[,<span class="dv">1</span>],</span>
<span id="cb19-6"><a href="capítulo-n-5---clustering.html#cb19-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y =</span> Xmat[,<span class="dv">2</span>],</span>
<span id="cb19-7"><a href="capítulo-n-5---clustering.html#cb19-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster =</span> <span class="fu">as.factor</span>(out1<span class="sc">$</span>cluster))</span>
<span id="cb19-8"><a href="capítulo-n-5---clustering.html#cb19-8" aria-hidden="true" tabindex="-1"></a>data2 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">x =</span> Xmat[,<span class="dv">1</span>],</span>
<span id="cb19-9"><a href="capítulo-n-5---clustering.html#cb19-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y =</span> Xmat[,<span class="dv">2</span>],</span>
<span id="cb19-10"><a href="capítulo-n-5---clustering.html#cb19-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster =</span> <span class="fu">as.factor</span>(out2<span class="sc">$</span>cluster))</span>
<span id="cb19-11"><a href="capítulo-n-5---clustering.html#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="capítulo-n-5---clustering.html#cb19-12" aria-hidden="true" tabindex="-1"></a>.mp <span class="ot">=</span> <span class="cf">function</span>(v, cdg) {</span>
<span id="cb19-13"><a href="capítulo-n-5---clustering.html#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> v, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb19-14"><a href="capítulo-n-5---clustering.html#cb19-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">col =</span> cluster, <span class="at">shape =</span> cluster), <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb19-15"><a href="capítulo-n-5---clustering.html#cb19-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> cdg, <span class="at">fill =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="dv">7</span>, <span class="at">shape =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb19-16"><a href="capítulo-n-5---clustering.html#cb19-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_discrete</span>(<span class="at">solid =</span> <span class="cn">TRUE</span>, <span class="at">guide =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">col =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb19-17"><a href="capítulo-n-5---clustering.html#cb19-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-18"><a href="capítulo-n-5---clustering.html#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="co"># centers of clusters:</span></span>
<span id="cb19-19"><a href="capítulo-n-5---clustering.html#cb19-19" aria-hidden="true" tabindex="-1"></a>cdg <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">x =</span> cents[,<span class="dv">1</span>],<span class="at">y =</span> cents[,<span class="dv">2</span>])</span>
<span id="cb19-20"><a href="capítulo-n-5---clustering.html#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="fu">.mp</span>(data0, cdg)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="capítulo-n-5---clustering.html#cb20-1" aria-hidden="true" tabindex="-1"></a>cents <span class="ot">=</span> out1<span class="sc">$</span>centers</span>
<span id="cb20-2"><a href="capítulo-n-5---clustering.html#cb20-2" aria-hidden="true" tabindex="-1"></a>cdg1 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">x=</span>cents[,<span class="dv">1</span>],<span class="at">y=</span>cents[,<span class="dv">2</span>])</span>
<span id="cb20-3"><a href="capítulo-n-5---clustering.html#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">.mp</span>(data1, cdg1)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-7-2.png" width="672" /></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="capítulo-n-5---clustering.html#cb21-1" aria-hidden="true" tabindex="-1"></a>cents <span class="ot">=</span> out2<span class="sc">$</span>centers</span>
<span id="cb21-2"><a href="capítulo-n-5---clustering.html#cb21-2" aria-hidden="true" tabindex="-1"></a>cdg2 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">x=</span>cents[,<span class="dv">1</span>],<span class="at">y=</span>cents[,<span class="dv">2</span>])</span>
<span id="cb21-3"><a href="capítulo-n-5---clustering.html#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">.mp</span>(data2, cdg2)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-7-3.png" width="672" /></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="capítulo-n-5---clustering.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(scRNAseq)</span>
<span id="cb22-2"><a href="capítulo-n-5---clustering.html#cb22-2" aria-hidden="true" tabindex="-1"></a>.oldMar <span class="ot">=</span> <span class="fu">par</span>(<span class="st">&quot;mar&quot;</span>)</span>
<span id="cb22-3"><a href="capítulo-n-5---clustering.html#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="fl">1.1</span>, <span class="dv">6</span>, <span class="fl">4.1</span>, <span class="fl">1.1</span>))</span>
<span id="cb22-4"><a href="capítulo-n-5---clustering.html#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;clusterExperiment&quot;</span>)</span>
<span id="cb22-5"><a href="capítulo-n-5---clustering.html#cb22-5" aria-hidden="true" tabindex="-1"></a>fluidigm <span class="ot">=</span> scRNAseq<span class="sc">::</span><span class="fu">ReprocessedFluidigmData</span>()</span>
<span id="cb22-6"><a href="capítulo-n-5---clustering.html#cb22-6" aria-hidden="true" tabindex="-1"></a>se <span class="ot">=</span> fluidigm[, fluidigm<span class="sc">$</span>Coverage_Type <span class="sc">==</span> <span class="st">&quot;High&quot;</span>]</span>
<span id="cb22-7"><a href="capítulo-n-5---clustering.html#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se) <span class="ot">=</span> <span class="fu">list</span>(<span class="at">normalized_counts =</span> </span>
<span id="cb22-8"><a href="capítulo-n-5---clustering.html#cb22-8" aria-hidden="true" tabindex="-1"></a>   <span class="fu">round</span>(limma<span class="sc">::</span><span class="fu">normalizeQuantiles</span>(<span class="fu">assay</span>(se))))</span>
<span id="cb22-9"><a href="capítulo-n-5---clustering.html#cb22-9" aria-hidden="true" tabindex="-1"></a>ce <span class="ot">=</span> <span class="fu">clusterMany</span>(se, <span class="at">clusterFunction =</span> <span class="st">&quot;pam&quot;</span>, <span class="at">ks =</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">run =</span> <span class="cn">TRUE</span>,</span>
<span id="cb22-10"><a href="capítulo-n-5---clustering.html#cb22-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">isCount =</span> <span class="cn">TRUE</span>, <span class="at">reduceMethod =</span> <span class="st">&quot;var&quot;</span>, <span class="at">nFilterDims =</span> <span class="fu">c</span>(<span class="dv">60</span>, <span class="dv">100</span>, <span class="dv">150</span>))</span></code></pre></div>
<pre><code>## 18 parameter combinations, 0 use sequential method, 0 use subsampling method
## Running Clustering on Parameter Combinations...
## done.</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="capítulo-n-5---clustering.html#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">clusterLabels</span>(ce) <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">&quot;FilterDims&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">clusterLabels</span>(ce))</span>
<span id="cb24-2"><a href="capítulo-n-5---clustering.html#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotClusters</span>(ce, <span class="at">whichClusters =</span> <span class="st">&quot;workflow&quot;</span>, <span class="at">axisLine =</span> <span class="sc">-</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="capítulo-n-5---clustering.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> .oldMar)</span></code></pre></div>
</div>
</div>
<div id="ejemplos-de-agrupamiento-citometría-de-flujo-y-citometría-de-masas" class="section level2 hasAnchor" number="5.5">
<h2><span class="header-section-number">5.5</span> Ejemplos de agrupamiento: citometría de flujo y citometría de masas<a href="capítulo-n-5---clustering.html#ejemplos-de-agrupamiento-citometría-de-flujo-y-citometría-de-masas" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="citometría-de-flujo-y-citometría-de-masas" class="section level3 hasAnchor" number="5.5.1">
<h3><span class="header-section-number">5.5.1</span> Citometría de flujo y citometría de masas<a href="capítulo-n-5---clustering.html#citometría-de-flujo-y-citometría-de-masas" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="capítulo-n-5---clustering.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;flowCore&quot;</span>,<span class="st">&quot;flowViz&quot;</span>)</span>
<span id="cb26-2"><a href="capítulo-n-5---clustering.html#cb26-2" aria-hidden="true" tabindex="-1"></a>fcsB <span class="ot">=</span> <span class="fu">read.FCS</span>(<span class="st">&quot;../data/Bendall_2011.fcs&quot;</span>, <span class="at">truncate_max_range =</span> <span class="cn">FALSE</span>)</span>
<span id="cb26-3"><a href="capítulo-n-5---clustering.html#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">slotNames</span>(fcsB)</span></code></pre></div>
<pre><code>## [1] &quot;exprs&quot;       &quot;parameters&quot;  &quot;description&quot;</code></pre>
</div>
<div id="preprocesamiento-de-datos" class="section level3 hasAnchor" number="5.5.2">
<h3><span class="header-section-number">5.5.2</span> Preprocesamiento de datos<a href="capítulo-n-5---clustering.html#preprocesamiento-de-datos" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="capítulo-n-5---clustering.html#cb28-1" aria-hidden="true" tabindex="-1"></a>markersB <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">&quot;../data/Bendall_2011_markers.csv&quot;</span>)</span></code></pre></div>
<pre><code>## Rows: 10 Columns: 2
## -- Column specification --------------------------------------------------------
## Delimiter: &quot;,&quot;
## chr (2): isotope, marker
## 
## i Use `spec()` to retrieve the full column specification for this data.
## i Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="capítulo-n-5---clustering.html#cb30-1" aria-hidden="true" tabindex="-1"></a>mt <span class="ot">=</span> <span class="fu">match</span>(markersB<span class="sc">$</span>isotope, <span class="fu">colnames</span>(fcsB))</span>
<span id="cb30-2"><a href="capítulo-n-5---clustering.html#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="sc">!</span><span class="fu">any</span>(<span class="fu">is.na</span>(mt)))</span>
<span id="cb30-3"><a href="capítulo-n-5---clustering.html#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(fcsB)[mt] <span class="ot">=</span> markersB<span class="sc">$</span>marker</span></code></pre></div>
<p>Mediciones de celda que muestran un agrupamiento claro en dos dimensiones.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="capítulo-n-5---clustering.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">flowPlot</span>(fcsB, <span class="at">plotParameters =</span> <span class="fu">colnames</span>(fcsB)[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">logy =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>Intente ejecutar el siguiente código para ver los dos regímenes principales de la transformación: valores pequeños y valores grandes.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="capítulo-n-5---clustering.html#cb32-1" aria-hidden="true" tabindex="-1"></a>v1 <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">100</span>)</span>
<span id="cb32-2"><a href="capítulo-n-5---clustering.html#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">log</span>(v1), <span class="fu">asinh</span>(v1), <span class="at">type =</span> <span class="st">&#39;l&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="capítulo-n-5---clustering.html#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(v1, <span class="fu">asinh</span>(v1), <span class="at">type =</span> <span class="st">&#39;l&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-12-2.png" width="672" /></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="capítulo-n-5---clustering.html#cb34-1" aria-hidden="true" tabindex="-1"></a>v3 <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">30</span>, <span class="dv">3000</span>, <span class="at">length =</span> <span class="dv">100</span>)</span>
<span id="cb34-2"><a href="capítulo-n-5---clustering.html#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">log</span>(v3), <span class="fu">asinh</span>(v3), <span class="at">type=</span> <span class="st">&#39;l&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-12-3.png" width="672" /></p>
<p>Este es otro ejemplo de una transformación estabilizadora de varianza.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="capítulo-n-5---clustering.html#cb35-1" aria-hidden="true" tabindex="-1"></a>asinhtrsf <span class="ot">=</span> <span class="fu">arcsinhTransform</span>(<span class="at">a =</span> <span class="fl">0.1</span>, <span class="at">b =</span> <span class="dv">1</span>)</span>
<span id="cb35-2"><a href="capítulo-n-5---clustering.html#cb35-2" aria-hidden="true" tabindex="-1"></a>fcsBT <span class="ot">=</span> <span class="fu">transform</span>(fcsB,</span>
<span id="cb35-3"><a href="capítulo-n-5---clustering.html#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transformList</span>(<span class="fu">colnames</span>(fcsB)[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">41</span>)], asinhtrsf))</span>
<span id="cb35-4"><a href="capítulo-n-5---clustering.html#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">densityplot</span>( <span class="sc">~</span><span class="st">`</span><span class="at">CD3all</span><span class="st">`</span>, fcsB)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="capítulo-n-5---clustering.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">densityplot</span>( <span class="sc">~</span><span class="st">`</span><span class="at">CD3all</span><span class="st">`</span>, fcsBT)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-13-2.png" width="672" /></p>
<p><strong>Pregunta</strong></p>
<p>¿Cuántas dimensiones usa el siguiente código para dividir los datos en 2 grupos usando k
-medio ?</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="capítulo-n-5---clustering.html#cb37-1" aria-hidden="true" tabindex="-1"></a>kf <span class="ot">=</span> <span class="fu">kmeansFilter</span>(<span class="st">&quot;CD3all&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;Pop1&quot;</span>,<span class="st">&quot;Pop2&quot;</span>), <span class="at">filterId=</span><span class="st">&quot;myKmFilter&quot;</span>)</span>
<span id="cb37-2"><a href="capítulo-n-5---clustering.html#cb37-2" aria-hidden="true" tabindex="-1"></a>fres <span class="ot">=</span> flowCore<span class="sc">::</span><span class="fu">filter</span>(fcsBT, kf)</span>
<span id="cb37-3"><a href="capítulo-n-5---clustering.html#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fres)</span></code></pre></div>
<pre><code>## Pop1: 33434 of 91392 events (36.58%)
## Pop2: 57958 of 91392 events (63.42%)</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="capítulo-n-5---clustering.html#cb39-1" aria-hidden="true" tabindex="-1"></a>fcsBT1 <span class="ot">=</span> flowCore<span class="sc">::</span><span class="fu">split</span>(fcsBT, fres, <span class="at">population =</span> <span class="st">&quot;Pop1&quot;</span>)</span>
<span id="cb39-2"><a href="capítulo-n-5---clustering.html#cb39-2" aria-hidden="true" tabindex="-1"></a>fcsBT2 <span class="ot">=</span> flowCore<span class="sc">::</span><span class="fu">split</span>(fcsBT, fres, <span class="at">population =</span> <span class="st">&quot;Pop2&quot;</span>)</span></code></pre></div>
<p>Después de la transformación, estas células se agruparon usando kmeans.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="capítulo-n-5---clustering.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;flowPeaks&quot;</span>)</span>
<span id="cb40-2"><a href="capítulo-n-5---clustering.html#cb40-2" aria-hidden="true" tabindex="-1"></a>fp <span class="ot">=</span> <span class="fu">flowPeaks</span>(Biobase<span class="sc">::</span><span class="fu">exprs</span>(fcsBT)[, <span class="fu">c</span>(<span class="st">&quot;CD3all&quot;</span>, <span class="st">&quot;CD56&quot;</span>)])</span></code></pre></div>
<pre><code>##         step 0, set the intial seeds, tot.wss=14636.9
##         step 1, do the rough EM, tot.wss=10278.9 at 0.513 sec
##         step 2, do the fine transfer of Hartigan-Wong Algorithm
##                  tot.wss=10221.1 at 1.294 sec</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="capítulo-n-5---clustering.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fp)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>Al graficar puntos que pueblan densamente un área, aquí se usa contornos y sombreado. Esto se hace de la siguiente manera:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="capítulo-n-5---clustering.html#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">flowPlot</span>(fcsBT, <span class="at">plotParameters =</span> <span class="fu">c</span>(<span class="st">&quot;CD3all&quot;</span>, <span class="st">&quot;CD56&quot;</span>), <span class="at">logy =</span> <span class="cn">FALSE</span>)</span>
<span id="cb43-2"><a href="capítulo-n-5---clustering.html#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">contour</span>(fcsBT[, <span class="fu">c</span>(<span class="dv">40</span>, <span class="dv">19</span>)], <span class="at">add =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="capítulo-n-5---clustering.html#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;SamSPECTRAL&quot;</span>)</span>
<span id="cb44-2"><a href="capítulo-n-5---clustering.html#cb44-2" aria-hidden="true" tabindex="-1"></a>mc2 <span class="ot">=</span> fcsBT<span class="sc">@</span>exprs[,<span class="fu">c</span>(<span class="dv">40</span>,<span class="dv">33</span>)]</span>
<span id="cb44-3"><a href="capítulo-n-5---clustering.html#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mc2)</span></code></pre></div>
<pre><code>##         CD3all      CD56
## [1,]  4.002830  2.144493
## [2,] -2.318107  1.272534
## [3,]  3.688635 -1.207639
## [4,]  1.440554 -1.530669
## [5,]  3.266283 -1.139235
## [6,]  3.023840  1.055028</code></pre>
<p>Intente comparar la salida usando este enfoque con lo que hicimos anteriormente usando lo siguiente:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="capítulo-n-5---clustering.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;ggcyto&quot;</span>,<span class="st">&quot;labeling&quot;</span>)</span>
<span id="cb46-2"><a href="capítulo-n-5---clustering.html#cb46-2" aria-hidden="true" tabindex="-1"></a>ggcd4cd8<span class="ot">=</span><span class="fu">ggcyto</span>(fcsB,<span class="fu">aes</span>(<span class="at">x=</span>CD4,<span class="at">y=</span>CD8))</span>
<span id="cb46-3"><a href="capítulo-n-5---clustering.html#cb46-3" aria-hidden="true" tabindex="-1"></a>ggcd4<span class="ot">=</span><span class="fu">ggcyto</span>(fcsB,<span class="fu">aes</span>(<span class="at">x=</span>CD4))</span>
<span id="cb46-4"><a href="capítulo-n-5---clustering.html#cb46-4" aria-hidden="true" tabindex="-1"></a>ggcd8<span class="ot">=</span><span class="fu">ggcyto</span>(fcsB,<span class="fu">aes</span>(<span class="at">x=</span>CD8))</span>
<span id="cb46-5"><a href="capítulo-n-5---clustering.html#cb46-5" aria-hidden="true" tabindex="-1"></a>p1<span class="ot">=</span>ggcd4<span class="sc">+</span><span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">60</span>)</span>
<span id="cb46-6"><a href="capítulo-n-5---clustering.html#cb46-6" aria-hidden="true" tabindex="-1"></a>p1b<span class="ot">=</span>ggcd8<span class="sc">+</span><span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">60</span>)</span>
<span id="cb46-7"><a href="capítulo-n-5---clustering.html#cb46-7" aria-hidden="true" tabindex="-1"></a>asinhT <span class="ot">=</span> <span class="fu">arcsinhTransform</span>(<span class="at">a=</span><span class="dv">0</span>,<span class="at">b=</span><span class="dv">1</span>)</span>
<span id="cb46-8"><a href="capítulo-n-5---clustering.html#cb46-8" aria-hidden="true" tabindex="-1"></a>transl <span class="ot">=</span> <span class="fu">transformList</span>(<span class="fu">colnames</span>(fcsB)[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">41</span>)], asinhT)</span>
<span id="cb46-9"><a href="capítulo-n-5---clustering.html#cb46-9" aria-hidden="true" tabindex="-1"></a>fcsBT <span class="ot">=</span> <span class="fu">transform</span>(fcsB, transl)</span>
<span id="cb46-10"><a href="capítulo-n-5---clustering.html#cb46-10" aria-hidden="true" tabindex="-1"></a>p1t<span class="ot">=</span><span class="fu">ggcyto</span>(fcsBT,<span class="fu">aes</span>(<span class="at">x=</span>CD4))<span class="sc">+</span><span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">90</span>)</span>
<span id="cb46-11"><a href="capítulo-n-5---clustering.html#cb46-11" aria-hidden="true" tabindex="-1"></a>p2t<span class="ot">=</span><span class="fu">ggcyto</span>(fcsBT,<span class="fu">aes</span>(<span class="at">x=</span>CD4,<span class="at">y=</span>CD8))<span class="sc">+</span><span class="fu">geom_density2d</span>(<span class="at">colour=</span><span class="st">&quot;black&quot;</span>)</span>
<span id="cb46-12"><a href="capítulo-n-5---clustering.html#cb46-12" aria-hidden="true" tabindex="-1"></a>p3t<span class="ot">=</span><span class="fu">ggcyto</span>(fcsBT,<span class="fu">aes</span>(<span class="at">x=</span>CD45RA,<span class="at">y=</span>CD20))<span class="sc">+</span><span class="fu">geom_density2d</span>(<span class="at">colour=</span><span class="st">&quot;black&quot;</span>)</span></code></pre></div>
</div>
<div id="agrupamiento-basado-en-la-densidad" class="section level3 hasAnchor" number="5.5.3">
<h3><span class="header-section-number">5.5.3</span> Agrupamiento basado en la densidad<a href="capítulo-n-5---clustering.html#agrupamiento-basado-en-la-densidad" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="capítulo-n-5---clustering.html#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;dbscan&quot;</span>)</span>
<span id="cb47-2"><a href="capítulo-n-5---clustering.html#cb47-2" aria-hidden="true" tabindex="-1"></a>mc5 <span class="ot">=</span> Biobase<span class="sc">::</span><span class="fu">exprs</span>(fcsBT)[, <span class="fu">c</span>(<span class="dv">15</span>,<span class="dv">16</span>,<span class="dv">19</span>,<span class="dv">40</span>,<span class="dv">33</span>)]</span>
<span id="cb47-3"><a href="capítulo-n-5---clustering.html#cb47-3" aria-hidden="true" tabindex="-1"></a>res5 <span class="ot">=</span> dbscan<span class="sc">::</span><span class="fu">dbscan</span>(mc5, <span class="at">eps =</span> <span class="fl">0.65</span>, <span class="at">minPts =</span> <span class="dv">30</span>)</span>
<span id="cb47-4"><a href="capítulo-n-5---clustering.html#cb47-4" aria-hidden="true" tabindex="-1"></a>mc5df <span class="ot">=</span> <span class="fu">data.frame</span>(mc5, <span class="at">cluster =</span> <span class="fu">as.factor</span>(res5<span class="sc">$</span>cluster))</span>
<span id="cb47-5"><a href="capítulo-n-5---clustering.html#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(mc5df<span class="sc">$</span>cluster)</span></code></pre></div>
<pre><code>## 
##     0     1     2     3     4     5     6     7     8 
## 76053  4031  5450  5310   257   160    63    25    43</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="capítulo-n-5---clustering.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mc5df, <span class="fu">aes</span>(<span class="at">x=</span>CD4,    <span class="at">y=</span>CD8,  <span class="at">col=</span>cluster))<span class="sc">+</span><span class="fu">geom_density2d</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="capítulo-n-5---clustering.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mc5df, <span class="fu">aes</span>(<span class="at">x=</span>CD3all, <span class="at">y=</span>CD20, <span class="at">col=</span>cluster))<span class="sc">+</span><span class="fu">geom_density2d</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-19-2.png" width="672" /></p>
<p>Estas dos gráficas muestran los resultados de agrupar con dbscanutilizando cinco marcadores. Aquí solo mostramos las proyecciones de los datos en los planos CD4-CD8 y C3all-CD20.</p>
</div>
</div>
<div id="agrupamiento-jerárquico" class="section level2 hasAnchor" number="5.6">
<h2><span class="header-section-number">5.6</span> Agrupamiento jerárquico<a href="capítulo-n-5---clustering.html#agrupamiento-jerárquico" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="cómo-calcular-des-similitudes-entre-clústeres-agregados" class="section level3 hasAnchor" number="5.6.1">
<h3><span class="header-section-number">5.6.1</span> ¿Cómo calcular (des) similitudes entre clústeres agregados?<a href="capítulo-n-5---clustering.html#cómo-calcular-des-similitudes-entre-clústeres-agregados" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="capítulo-n-5---clustering.html#cb51-1" aria-hidden="true" tabindex="-1"></a>mc6 <span class="ot">=</span> Biobase<span class="sc">::</span><span class="fu">exprs</span>(fcsBT)[, <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">19</span>, <span class="dv">33</span>, <span class="dv">25</span>, <span class="dv">40</span>)]</span>
<span id="cb51-2"><a href="capítulo-n-5---clustering.html#cb51-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> dbscan<span class="sc">::</span><span class="fu">dbscan</span>(mc6, <span class="at">eps =</span> <span class="fl">0.65</span>, <span class="at">minPts =</span> <span class="dv">20</span>)</span>
<span id="cb51-3"><a href="capítulo-n-5---clustering.html#cb51-3" aria-hidden="true" tabindex="-1"></a>mc6df <span class="ot">=</span> <span class="fu">data.frame</span>(mc6, <span class="at">cluster =</span> <span class="fu">as.factor</span>(res<span class="sc">$</span>cluster))</span>
<span id="cb51-4"><a href="capítulo-n-5---clustering.html#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(mc6df<span class="sc">$</span>cluster)</span></code></pre></div>
<pre><code>## 
##     0     1     2     3     4     5     6 
## 91068    34    61    20    67   121    21</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="capítulo-n-5---clustering.html#cb53-1" aria-hidden="true" tabindex="-1"></a>mc7 <span class="ot">=</span> Biobase<span class="sc">::</span><span class="fu">exprs</span>(fcsBT)[, <span class="fu">c</span>(<span class="dv">11</span>, <span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">19</span>, <span class="dv">25</span>, <span class="dv">33</span>, <span class="dv">40</span>)]</span>
<span id="cb53-2"><a href="capítulo-n-5---clustering.html#cb53-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> dbscan<span class="sc">::</span><span class="fu">dbscan</span>(mc7, <span class="at">eps =</span> <span class="fl">0.95</span>, <span class="at">minPts =</span> <span class="dv">20</span>)</span>
<span id="cb53-3"><a href="capítulo-n-5---clustering.html#cb53-3" aria-hidden="true" tabindex="-1"></a>mc7df <span class="ot">=</span> <span class="fu">data.frame</span>(mc7, <span class="at">cluster =</span> <span class="fu">as.factor</span>(res<span class="sc">$</span>cluster))</span>
<span id="cb53-4"><a href="capítulo-n-5---clustering.html#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(mc7df<span class="sc">$</span>cluster)</span></code></pre></div>
<pre><code>## 
##     0     1     2     3     4     5     6     7     8     9    10 
## 90249    21   102   445   158   119    19   224    17    20    18</code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="capítulo-n-5---clustering.html#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="do">## ####trees together par(mfrow)) not working for pheatmap</span></span>
<span id="cb55-2"><a href="capítulo-n-5---clustering.html#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="do">## library(&quot;pheatmap&quot;)</span></span>
<span id="cb55-3"><a href="capítulo-n-5---clustering.html#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="do">## load(&quot;../data/d14.RData&quot;)</span></span>
<span id="cb55-4"><a href="capítulo-n-5---clustering.html#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="do">## pheatmap(d14,clustering_distance_rows=d14,treeheight_col =200,</span></span>
<span id="cb55-5"><a href="capítulo-n-5---clustering.html#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="do">## cellwidth=20,cellheight=10,lwd=5,treeheight_row=0,clustering_method = &quot;single&quot;,</span></span>
<span id="cb55-6"><a href="capítulo-n-5---clustering.html#cb55-6" aria-hidden="true" tabindex="-1"></a><span class="do">## labels_col=1:11,main=&quot;single&quot;)</span></span>
<span id="cb55-7"><a href="capítulo-n-5---clustering.html#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="do">## pheatmap(d14,clustering_distance_rows=d14,treeheight_col =200,cellwidth=20,</span></span>
<span id="cb55-8"><a href="capítulo-n-5---clustering.html#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="do">## cellheight=10,lwd=5,treeheight_row=0,clustering_method = &quot;average&quot;,</span></span>
<span id="cb55-9"><a href="capítulo-n-5---clustering.html#cb55-9" aria-hidden="true" tabindex="-1"></a><span class="do">## labels_col=1:11,main=&quot;average&quot;)</span></span>
<span id="cb55-10"><a href="capítulo-n-5---clustering.html#cb55-10" aria-hidden="true" tabindex="-1"></a><span class="do">## pheatmap(d14,clustering_distance_rows=d14,treeheight_col =200,cellwidth=20,</span></span>
<span id="cb55-11"><a href="capítulo-n-5---clustering.html#cb55-11" aria-hidden="true" tabindex="-1"></a><span class="do">## cellheight=10,lwd=5,treeheight_row=0,clustering_method = &quot;complete&quot;,labels_col=1:11,</span></span>
<span id="cb55-12"><a href="capítulo-n-5---clustering.html#cb55-12" aria-hidden="true" tabindex="-1"></a><span class="do">## main=&quot;complete&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="capítulo-n-5---clustering.html#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="do">## load(&quot;../data/Msig3transp.RData&quot;)</span></span>
<span id="cb56-2"><a href="capítulo-n-5---clustering.html#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="do">## celltypes=factor(substr(rownames(Msig3transp),7,9))</span></span>
<span id="cb56-3"><a href="capítulo-n-5---clustering.html#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="do">## status=factor(substr(rownames(Msig3transp),1,3))</span></span>
<span id="cb56-4"><a href="capítulo-n-5---clustering.html#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Msig2=as.matrix(Msig3transp)</span></span>
<span id="cb56-5"><a href="capítulo-n-5---clustering.html#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="do">## rownames(Msig2)=substr(rownames(Msig2),1,9)</span></span>
<span id="cb56-6"><a href="capítulo-n-5---clustering.html#cb56-6" aria-hidden="true" tabindex="-1"></a><span class="do">## hm1=heatmap(as.matrix(dist(Msig2)))</span></span>
<span id="cb56-7"><a href="capítulo-n-5---clustering.html#cb56-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Morder=Msig2[hm1$rowInd,]</span></span>
<span id="cb56-8"><a href="capítulo-n-5---clustering.html#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="do">## save(Morder,file=&quot;../data/Morder.RData&quot;)</span></span>
<span id="cb56-9"><a href="capítulo-n-5---clustering.html#cb56-9" aria-hidden="true" tabindex="-1"></a><span class="do">## write.table(Morder,&quot;../data/Morder.txt&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="capítulo-n-5---clustering.html#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># p_load(&quot;gplots&quot;,&quot;pheatmap&quot;,&quot;RColorBrewer&quot;)</span></span>
<span id="cb57-2"><a href="capítulo-n-5---clustering.html#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load(&quot;../data/Morder.RData&quot;)</span></span>
<span id="cb57-3"><a href="capítulo-n-5---clustering.html#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co"># celltypes=factor(substr(rownames(Morder),7,9))</span></span>
<span id="cb57-4"><a href="capítulo-n-5---clustering.html#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co"># status=factor(substr(rownames(Morder),1,3))</span></span>
<span id="cb57-5"><a href="capítulo-n-5---clustering.html#cb57-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ##Just the Euclidean distance</span></span>
<span id="cb57-6"><a href="capítulo-n-5---clustering.html#cb57-6" aria-hidden="true" tabindex="-1"></a><span class="co"># pheatmap(as.matrix(dist(Morder)),cluster_rows=FALSE,</span></span>
<span id="cb57-7"><a href="capítulo-n-5---clustering.html#cb57-7" aria-hidden="true" tabindex="-1"></a><span class="co">#         cluster_cols=FALSE,cellwidth=10,cellheight=10)</span></span>
<span id="cb57-8"><a href="capítulo-n-5---clustering.html#cb57-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ###Manhattan</span></span>
<span id="cb57-9"><a href="capítulo-n-5---clustering.html#cb57-9" aria-hidden="true" tabindex="-1"></a><span class="co"># pheatmap(as.matrix(dist(Morder,&quot;manhattan&quot;)),cluster_rows=FALSE,</span></span>
<span id="cb57-10"><a href="capítulo-n-5---clustering.html#cb57-10" aria-hidden="true" tabindex="-1"></a><span class="co">#         cluster_cols=FALSE,cellwidth=10,cellheight=10)</span></span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="capítulo-n-5---clustering.html#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pheatmap(corT,clustering_distance_rows=distcor,</span></span>
<span id="cb58-2"><a href="capítulo-n-5---clustering.html#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="co">#        annotation_row=samplesdata[,c(&quot;celltypes&quot;,&quot;status&quot;)],</span></span>
<span id="cb58-3"><a href="capítulo-n-5---clustering.html#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="co">#        show_rownames = FALSE, show_colnames = FALSE)</span></span>
<span id="cb58-4"><a href="capítulo-n-5---clustering.html#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="co"># pheatmap(corT,clustering_distance_rows=distcor,treeheight_row =150,</span></span>
<span id="cb58-5"><a href="capítulo-n-5---clustering.html#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="co">#        annotation_row=samplesdata[,c(&quot;celltypes&quot;,&quot;status&quot;)],</span></span>
<span id="cb58-6"><a href="capítulo-n-5---clustering.html#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="co">#        show_rownames = FALSE, show_colnames = FALSE)</span></span>
<span id="cb58-7"><a href="capítulo-n-5---clustering.html#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="co"># pheatmap(corT,clustering_distance_rows=distcor,treeheight_row =150,</span></span>
<span id="cb58-8"><a href="capítulo-n-5---clustering.html#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="co">#        annotation_row=samplesdata[,c(&quot;celltypes&quot;,&quot;status&quot;)],</span></span>
<span id="cb58-9"><a href="capítulo-n-5---clustering.html#cb58-9" aria-hidden="true" tabindex="-1"></a><span class="co">#        treeheight_col =150,</span></span>
<span id="cb58-10"><a href="capítulo-n-5---clustering.html#cb58-10" aria-hidden="true" tabindex="-1"></a><span class="co">#        show_rownames = FALSE, show_colnames = FALSE)</span></span>
<span id="cb58-11"><a href="capítulo-n-5---clustering.html#cb58-11" aria-hidden="true" tabindex="-1"></a><span class="co"># pheatmap(corT,clustering_distance_rows=distcor,treeheight_row =150,</span></span>
<span id="cb58-12"><a href="capítulo-n-5---clustering.html#cb58-12" aria-hidden="true" tabindex="-1"></a><span class="co">#           annotation_col=samplesdata[,c(&quot;celltypes&quot;,&quot;status&quot;)],</span></span>
<span id="cb58-13"><a href="capítulo-n-5---clustering.html#cb58-13" aria-hidden="true" tabindex="-1"></a><span class="co">#        annotation_row=samplesdata[,c(&quot;celltypes&quot;,&quot;status&quot;)],</span></span>
<span id="cb58-14"><a href="capítulo-n-5---clustering.html#cb58-14" aria-hidden="true" tabindex="-1"></a><span class="co">#        treeheight_col =150,</span></span>
<span id="cb58-15"><a href="capítulo-n-5---clustering.html#cb58-15" aria-hidden="true" tabindex="-1"></a><span class="co">#        show_rownames = FALSE, show_colnames = FALSE)</span></span></code></pre></div>
</div>
</div>
<div id="validación-y-elección-del-número-de-clústeres" class="section level2 hasAnchor" number="5.7">
<h2><span class="header-section-number">5.7</span> Validación y elección del número de clústeres<a href="capítulo-n-5---clustering.html#validación-y-elección-del-número-de-clústeres" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="capítulo-n-5---clustering.html#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;dplyr&quot;</span>)</span>
<span id="cb59-2"><a href="capítulo-n-5---clustering.html#cb59-2" aria-hidden="true" tabindex="-1"></a>simdat <span class="ot">=</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>), <span class="cf">function</span>(mx) {</span>
<span id="cb59-3"><a href="capítulo-n-5---clustering.html#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>), <span class="cf">function</span>(my) {</span>
<span id="cb59-4"><a href="capítulo-n-5---clustering.html#cb59-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(<span class="at">x =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="at">mean =</span> mx, <span class="at">sd =</span> <span class="dv">2</span>),</span>
<span id="cb59-5"><a href="capítulo-n-5---clustering.html#cb59-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="at">mean =</span> my, <span class="at">sd =</span> <span class="dv">2</span>),</span>
<span id="cb59-6"><a href="capítulo-n-5---clustering.html#cb59-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">class =</span> <span class="fu">paste</span>(mx, my, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>))</span>
<span id="cb59-7"><a href="capítulo-n-5---clustering.html#cb59-7" aria-hidden="true" tabindex="-1"></a>   }) <span class="sc">%&gt;%</span> bind_rows</span>
<span id="cb59-8"><a href="capítulo-n-5---clustering.html#cb59-8" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> bind_rows</span>
<span id="cb59-9"><a href="capítulo-n-5---clustering.html#cb59-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(simdat)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 3
##        x      y class
##    &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;
## 1 -2.42  -4.59  0:0  
## 2  1.89  -1.56  0:0  
## 3  0.558  2.17  0:0  
## 4  2.51  -0.873 0:0  
## 5 -2.52  -0.766 0:0  
## 6  3.62   0.953 0:0</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="capítulo-n-5---clustering.html#cb61-1" aria-hidden="true" tabindex="-1"></a>simdatxy <span class="ot">=</span> simdat[, <span class="fu">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>)] </span></code></pre></div>
<p>Aquí, conocemos las etiquetas desde que generamos los datos; por lo general, no las conocemos.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="capítulo-n-5---clustering.html#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(simdat, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> class)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb62-2"><a href="capítulo-n-5---clustering.html#cb62-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-27-1.png" width="672" /></p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="capítulo-n-5---clustering.html#cb63-1" aria-hidden="true" tabindex="-1"></a>wss <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">k =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="at">value =</span> <span class="cn">NA_real_</span>)</span>
<span id="cb63-2"><a href="capítulo-n-5---clustering.html#cb63-2" aria-hidden="true" tabindex="-1"></a>wss<span class="sc">$</span>value[<span class="dv">1</span>] <span class="ot">=</span> <span class="fu">sum</span>(<span class="fu">scale</span>(simdatxy, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb63-3"><a href="capítulo-n-5---clustering.html#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">nrow</span>(wss)) {</span>
<span id="cb63-4"><a href="capítulo-n-5---clustering.html#cb63-4" aria-hidden="true" tabindex="-1"></a>  km  <span class="ot">=</span> <span class="fu">kmeans</span>(simdatxy, <span class="at">centers =</span> wss<span class="sc">$</span>k[i])</span>
<span id="cb63-5"><a href="capítulo-n-5---clustering.html#cb63-5" aria-hidden="true" tabindex="-1"></a>  wss<span class="sc">$</span>value[i] <span class="ot">=</span> <span class="fu">sum</span>(km<span class="sc">$</span>withinss)</span>
<span id="cb63-6"><a href="capítulo-n-5---clustering.html#cb63-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb63-7"><a href="capítulo-n-5---clustering.html#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(wss, <span class="fu">aes</span>(<span class="at">x =</span> k, <span class="at">y =</span> value)) <span class="sc">+</span> <span class="fu">geom_col</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<p>El gráfico de barras del estadístico WSS en función de k muestra que el último salto sustancial es justo antes k = 4 . Esto indica que la mejor opción para estos datos es k = 4 .</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="capítulo-n-5---clustering.html#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;fpc&quot;</span>,<span class="st">&quot;cluster&quot;</span>)</span>
<span id="cb64-2"><a href="capítulo-n-5---clustering.html#cb64-2" aria-hidden="true" tabindex="-1"></a>CH <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb64-3"><a href="capítulo-n-5---clustering.html#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">k =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">8</span>,</span>
<span id="cb64-4"><a href="capítulo-n-5---clustering.html#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">value =</span> <span class="fu">sapply</span>(k, <span class="cf">function</span>(i) {</span>
<span id="cb64-5"><a href="capítulo-n-5---clustering.html#cb64-5" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">=</span> <span class="fu">pam</span>(simdatxy, i)</span>
<span id="cb64-6"><a href="capítulo-n-5---clustering.html#cb64-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">calinhara</span>(simdatxy, p<span class="sc">$</span>cluster)</span>
<span id="cb64-7"><a href="capítulo-n-5---clustering.html#cb64-7" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb64-8"><a href="capítulo-n-5---clustering.html#cb64-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-9"><a href="capítulo-n-5---clustering.html#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(CH, <span class="fu">aes</span>(<span class="at">x =</span> k, <span class="at">y =</span> value)) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-10"><a href="capítulo-n-5---clustering.html#cb64-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;CH index&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-29-1.png" width="672" /></p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="capítulo-n-5---clustering.html#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;cluster&quot;</span>,<span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb65-2"><a href="capítulo-n-5---clustering.html#cb65-2" aria-hidden="true" tabindex="-1"></a>pamfun <span class="ot">=</span> <span class="cf">function</span>(x, k)</span>
<span id="cb65-3"><a href="capítulo-n-5---clustering.html#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="at">cluster =</span> <span class="fu">pam</span>(x, k, <span class="at">cluster.only =</span> <span class="cn">TRUE</span>))</span>
<span id="cb65-4"><a href="capítulo-n-5---clustering.html#cb65-4" aria-hidden="true" tabindex="-1"></a>gss <span class="ot">=</span> <span class="fu">clusGap</span>(simdatxy, <span class="at">FUN =</span> pamfun, <span class="at">K.max =</span> <span class="dv">8</span>, <span class="at">B =</span> <span class="dv">50</span>,</span>
<span id="cb65-5"><a href="capítulo-n-5---clustering.html#cb65-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb65-6"><a href="capítulo-n-5---clustering.html#cb65-6" aria-hidden="true" tabindex="-1"></a>plot_gap <span class="ot">=</span> <span class="cf">function</span>(x) {</span>
<span id="cb65-7"><a href="capítulo-n-5---clustering.html#cb65-7" aria-hidden="true" tabindex="-1"></a>  gstab <span class="ot">=</span> <span class="fu">data.frame</span>(x<span class="sc">$</span>Tab, <span class="at">k =</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(x<span class="sc">$</span>Tab)))</span>
<span id="cb65-8"><a href="capítulo-n-5---clustering.html#cb65-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(gstab, <span class="fu">aes</span>(k, gap)) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb65-9"><a href="capítulo-n-5---clustering.html#cb65-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymax =</span> gap <span class="sc">+</span> SE.sim,</span>
<span id="cb65-10"><a href="capítulo-n-5---clustering.html#cb65-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymin =</span> gap <span class="sc">-</span> SE.sim), <span class="at">width=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb65-11"><a href="capítulo-n-5---clustering.html#cb65-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>, <span class="at">col=</span>  <span class="st">&quot;red&quot;</span>)</span>
<span id="cb65-12"><a href="capítulo-n-5---clustering.html#cb65-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb65-13"><a href="capítulo-n-5---clustering.html#cb65-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_gap</span>(gss)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-30-1.png" width="672" /></p>
<div id="uso-de-la-estadística-de-la-brecha" class="section level3 hasAnchor" number="5.7.1">
<h3><span class="header-section-number">5.7.1</span> Uso de la estadística de la brecha<a href="capítulo-n-5---clustering.html#uso-de-la-estadística-de-la-brecha" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Haga una función que grafique la estadística de la brecha como en la Figura 5.27 . Mostrar la salida para el simdatconjunto de datos de ejemplo agrupado con el pamfunción.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="capítulo-n-5---clustering.html#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;Hiiragi2013&quot;</span>)</span>
<span id="cb66-2"><a href="capítulo-n-5---clustering.html#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;x&quot;</span>)</span></code></pre></div>
<p>Empezamos eligiendo los 50 genes (características) más variables.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="capítulo-n-5---clustering.html#cb67-1" aria-hidden="true" tabindex="-1"></a>selFeats <span class="ot">=</span> <span class="fu">order</span>(<span class="fu">rowVars</span>(Biobase<span class="sc">::</span><span class="fu">exprs</span>(x)), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>]</span>
<span id="cb67-2"><a href="capítulo-n-5---clustering.html#cb67-2" aria-hidden="true" tabindex="-1"></a>embmat <span class="ot">=</span> <span class="fu">t</span>(Biobase<span class="sc">::</span><span class="fu">exprs</span>(x)[selFeats, ])</span>
<span id="cb67-3"><a href="capítulo-n-5---clustering.html#cb67-3" aria-hidden="true" tabindex="-1"></a>embgap <span class="ot">=</span> <span class="fu">clusGap</span>(embmat, <span class="at">FUN =</span> pamfun, <span class="at">K.max =</span> <span class="dv">24</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb67-4"><a href="capítulo-n-5---clustering.html#cb67-4" aria-hidden="true" tabindex="-1"></a>k1 <span class="ot">=</span> <span class="fu">maxSE</span>(embgap<span class="sc">$</span>Tab[, <span class="st">&quot;gap&quot;</span>], embgap<span class="sc">$</span>Tab[, <span class="st">&quot;SE.sim&quot;</span>])</span>
<span id="cb67-5"><a href="capítulo-n-5---clustering.html#cb67-5" aria-hidden="true" tabindex="-1"></a>k2 <span class="ot">=</span> <span class="fu">maxSE</span>(embgap<span class="sc">$</span>Tab[, <span class="st">&quot;gap&quot;</span>], embgap<span class="sc">$</span>Tab[, <span class="st">&quot;SE.sim&quot;</span>],</span>
<span id="cb67-6"><a href="capítulo-n-5---clustering.html#cb67-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">method =</span> <span class="st">&quot;Tibs2001SEmax&quot;</span>)</span>
<span id="cb67-7"><a href="capítulo-n-5---clustering.html#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(k1, k2)</span></code></pre></div>
<pre><code>## [1] 9 7</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="capítulo-n-5---clustering.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="st">&quot;firstSEmax&quot;</span> <span class="sc">==</span> <span class="fu">eval</span>(<span class="fu">formals</span>(maxSE)<span class="sc">$</span>method)[<span class="dv">1</span>])</span></code></pre></div>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="capítulo-n-5---clustering.html#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(embgap, <span class="at">main =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="capítulo-n-5---clustering.html#cb71-1" aria-hidden="true" tabindex="-1"></a>cl <span class="ot">=</span> <span class="fu">pamfun</span>(embmat, <span class="at">k =</span> k1)<span class="sc">$</span>cluster</span>
<span id="cb71-2"><a href="capítulo-n-5---clustering.html#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">pData</span>(x)[<span class="fu">names</span>(cl), <span class="st">&quot;sampleGroup&quot;</span>], cl)</span></code></pre></div>
<pre><code>##                  cl
##                    1  2  3  4  5  6  7  8  9
##   E3.25           23 11  1  1  0  0  0  0  0
##   E3.25 (FGF4-KO)  0  0  1 16  0  0  0  0  0
##   E3.5 (EPI)       2  1  0  0  0  8  0  0  0
##   E3.5 (FGF4-KO)   0  0  8  0  0  0  0  0  0
##   E3.5 (PE)        0  0  0  0  9  2  0  0  0
##   E4.5 (EPI)       0  0  0  0  0  0  0  4  0
##   E4.5 (FGF4-KO)   0  0  0  0  0  0  0  0 10
##   E4.5 (PE)        0  0  0  0  0  0  4  0  0</code></pre>
<p>La estadística de brecha para los Hiiragi2013 .</p>
</div>
<div id="validación-de-clúster-mediante-bootstrap" class="section level3 hasAnchor" number="5.7.2">
<h3><span class="header-section-number">5.7.2</span> Validación de clúster mediante bootstrap<a href="capítulo-n-5---clustering.html#validación-de-clúster-mediante-bootstrap" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="capítulo-n-5---clustering.html#cb73-1" aria-hidden="true" tabindex="-1"></a>clusterResampling <span class="ot">=</span> <span class="cf">function</span>(x, <span class="at">ngenes =</span> <span class="dv">50</span>, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">B =</span> <span class="dv">250</span>,</span>
<span id="cb73-2"><a href="capítulo-n-5---clustering.html#cb73-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">prob =</span> <span class="fl">0.67</span>) {</span>
<span id="cb73-3"><a href="capítulo-n-5---clustering.html#cb73-3" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> Biobase<span class="sc">::</span><span class="fu">exprs</span>(x)</span>
<span id="cb73-4"><a href="capítulo-n-5---clustering.html#cb73-4" aria-hidden="true" tabindex="-1"></a>  ce <span class="ot">=</span> <span class="fu">cl_ensemble</span>(<span class="at">list =</span> <span class="fu">lapply</span>(<span class="fu">seq_len</span>(B), <span class="cf">function</span>(b) {</span>
<span id="cb73-5"><a href="capítulo-n-5---clustering.html#cb73-5" aria-hidden="true" tabindex="-1"></a>    selSamps <span class="ot">=</span> <span class="fu">sample</span>(<span class="fu">ncol</span>(mat), <span class="at">size =</span> <span class="fu">round</span>(prob <span class="sc">*</span> <span class="fu">ncol</span>(mat)),</span>
<span id="cb73-6"><a href="capítulo-n-5---clustering.html#cb73-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb73-7"><a href="capítulo-n-5---clustering.html#cb73-7" aria-hidden="true" tabindex="-1"></a>    submat <span class="ot">=</span> mat[, selSamps, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb73-8"><a href="capítulo-n-5---clustering.html#cb73-8" aria-hidden="true" tabindex="-1"></a>    sel <span class="ot">=</span> <span class="fu">order</span>(<span class="fu">rowVars</span>(submat), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="fu">seq_len</span>(ngenes)]</span>
<span id="cb73-9"><a href="capítulo-n-5---clustering.html#cb73-9" aria-hidden="true" tabindex="-1"></a>    submat <span class="ot">=</span> submat[sel,, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb73-10"><a href="capítulo-n-5---clustering.html#cb73-10" aria-hidden="true" tabindex="-1"></a>    pamres <span class="ot">=</span> <span class="fu">pam</span>(<span class="fu">t</span>(submat), <span class="at">k =</span> k)</span>
<span id="cb73-11"><a href="capítulo-n-5---clustering.html#cb73-11" aria-hidden="true" tabindex="-1"></a>    pred <span class="ot">=</span> <span class="fu">cl_predict</span>(pamres, <span class="fu">t</span>(mat[sel, ]), <span class="st">&quot;memberships&quot;</span>)</span>
<span id="cb73-12"><a href="capítulo-n-5---clustering.html#cb73-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.cl_partition</span>(pred)</span>
<span id="cb73-13"><a href="capítulo-n-5---clustering.html#cb73-13" aria-hidden="true" tabindex="-1"></a>  }))</span>
<span id="cb73-14"><a href="capítulo-n-5---clustering.html#cb73-14" aria-hidden="true" tabindex="-1"></a>  cons <span class="ot">=</span> <span class="fu">cl_consensus</span>(ce)</span>
<span id="cb73-15"><a href="capítulo-n-5---clustering.html#cb73-15" aria-hidden="true" tabindex="-1"></a>  ag <span class="ot">=</span> <span class="fu">sapply</span>(ce, cl_agreement, <span class="at">y =</span> cons)</span>
<span id="cb73-16"><a href="capítulo-n-5---clustering.html#cb73-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="at">agreements =</span> ag, <span class="at">consensus =</span> cons)</span>
<span id="cb73-17"><a href="capítulo-n-5---clustering.html#cb73-17" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="capítulo-n-5---clustering.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(sampling, clue)</span>
<span id="cb74-2"><a href="capítulo-n-5---clustering.html#cb74-2" aria-hidden="true" tabindex="-1"></a>iswt <span class="ot">=</span> (x<span class="sc">$</span>genotype <span class="sc">==</span> <span class="st">&quot;WT&quot;</span>)</span>
<span id="cb74-3"><a href="capítulo-n-5---clustering.html#cb74-3" aria-hidden="true" tabindex="-1"></a>cr1 <span class="ot">=</span> <span class="fu">clusterResampling</span>(x[, x<span class="sc">$</span>Embryonic.day <span class="sc">==</span> <span class="st">&quot;E3.25&quot;</span> <span class="sc">&amp;</span> iswt])</span>
<span id="cb74-4"><a href="capítulo-n-5---clustering.html#cb74-4" aria-hidden="true" tabindex="-1"></a>cr2 <span class="ot">=</span> <span class="fu">clusterResampling</span>(x[, x<span class="sc">$</span>Embryonic.day <span class="sc">==</span> <span class="st">&quot;E3.5&quot;</span>  <span class="sc">&amp;</span> iswt])</span></code></pre></div>
<p>Confirman la hipótesis de que los datos E.35 se dividen en dos grupos.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="capítulo-n-5---clustering.html#cb75-1" aria-hidden="true" tabindex="-1"></a>ag1 <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">agreements =</span> cr1<span class="sc">$</span>agreements, <span class="at">day =</span> <span class="st">&quot;E3.25&quot;</span>)</span>
<span id="cb75-2"><a href="capítulo-n-5---clustering.html#cb75-2" aria-hidden="true" tabindex="-1"></a>ag2 <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">agreements =</span> cr2<span class="sc">$</span>agreements, <span class="at">day =</span> <span class="st">&quot;E3.5&quot;</span>)</span>
<span id="cb75-3"><a href="capítulo-n-5---clustering.html#cb75-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">bind_rows</span>(ag1, ag2), <span class="fu">aes</span>(<span class="at">x =</span> day, <span class="at">y =</span> agreements)) <span class="sc">+</span></span>
<span id="cb75-4"><a href="capítulo-n-5---clustering.html#cb75-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb75-5"><a href="capítulo-n-5---clustering.html#cb75-5" aria-hidden="true" tabindex="-1"></a>  ggbeeswarm<span class="sc">::</span><span class="fu">geom_beeswarm</span>(<span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">col =</span> <span class="st">&quot;#0000ff40&quot;</span>)</span>
<span id="cb75-6"><a href="capítulo-n-5---clustering.html#cb75-6" aria-hidden="true" tabindex="-1"></a>mem1 <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">y =</span> <span class="fu">sort</span>(<span class="fu">cl_membership</span>(cr1<span class="sc">$</span>consensus)[, <span class="dv">1</span>]),</span>
<span id="cb75-7"><a href="capítulo-n-5---clustering.html#cb75-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">x =</span> <span class="fu">seq</span>(<span class="at">along =</span> y), <span class="at">day =</span> <span class="st">&quot;E3.25&quot;</span>)</span>
<span id="cb75-8"><a href="capítulo-n-5---clustering.html#cb75-8" aria-hidden="true" tabindex="-1"></a>mem2 <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">y =</span> <span class="fu">sort</span>(<span class="fu">cl_membership</span>(cr2<span class="sc">$</span>consensus)[, <span class="dv">1</span>]),</span>
<span id="cb75-9"><a href="capítulo-n-5---clustering.html#cb75-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">x =</span> <span class="fu">seq</span>(<span class="at">along =</span> y), <span class="at">day =</span> <span class="st">&quot;E3.5&quot;</span>)</span>
<span id="cb75-10"><a href="capítulo-n-5---clustering.html#cb75-10" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">bind_rows</span>(mem1, mem2), <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> day)) <span class="sc">+</span></span>
<span id="cb75-11"><a href="capítulo-n-5---clustering.html#cb75-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">facet_grid</span>(<span class="sc">~</span> day, <span class="at">scales =</span> <span class="st">&quot;free_x&quot;</span>)</span>
<span id="cb75-12"><a href="capítulo-n-5---clustering.html#cb75-12" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2, <span class="at">widths =</span> <span class="fu">c</span>(<span class="fl">2.4</span>,<span class="fl">4.0</span>))</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-37-1.png" width="672" /></p>
</div>
</div>
<div id="el-agrupamiento-como-medio-para-eliminar-el-ruido" class="section level2 hasAnchor" number="5.8">
<h2><span class="header-section-number">5.8</span> El agrupamiento como medio para eliminar el ruido<a href="capítulo-n-5---clustering.html#el-agrupamiento-como-medio-para-eliminar-el-ruido" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="observaciones-ruidosas-con-diferentes-frecuencias-de-línea-de-base" class="section level3 hasAnchor" number="5.8.1">
<h3><span class="header-section-number">5.8.1</span> Observaciones ruidosas con diferentes frecuencias de línea de base<a href="capítulo-n-5---clustering.html#observaciones-ruidosas-con-diferentes-frecuencias-de-línea-de-base" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="capítulo-n-5---clustering.html#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;mixtools&quot;</span>,<span class="st">&quot;ggplot2&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: package &#39;mixtools&#39; was built under R version 4.1.3</code></pre>
<pre><code>## mixtools package, version 1.2.0, Released 2020-02-05
## This package is based upon work supported by the National Science Foundation under Grant No. SES-0518772.</code></pre>
<pre><code>## 
## Attaching package: &#39;mixtools&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:gtools&#39;:
## 
##     ddirichlet</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="capítulo-n-5---clustering.html#cb81-1" aria-hidden="true" tabindex="-1"></a>seq1 <span class="ot">=</span> <span class="fu">rmvnorm</span>(<span class="at">n =</span> <span class="fl">1e3</span>, <span class="at">mu =</span> <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="at">sigma =</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">diag</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)))</span>
<span id="cb81-2"><a href="capítulo-n-5---clustering.html#cb81-2" aria-hidden="true" tabindex="-1"></a>seq2 <span class="ot">=</span> <span class="fu">rmvnorm</span>(<span class="at">n =</span> <span class="fl">1e5</span>, <span class="at">mu =</span>  <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="at">sigma =</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">diag</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)))</span>
<span id="cb81-3"><a href="capítulo-n-5---clustering.html#cb81-3" aria-hidden="true" tabindex="-1"></a>twogr <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb81-4"><a href="capítulo-n-5---clustering.html#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbind</span>(seq1, seq2),</span>
<span id="cb81-5"><a href="capítulo-n-5---clustering.html#cb81-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">seq =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(seq1)),</span>
<span id="cb81-6"><a href="capítulo-n-5---clustering.html#cb81-6" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">rep</span>(<span class="dv">2</span>, <span class="fu">nrow</span>(seq2))))</span>
<span id="cb81-7"><a href="capítulo-n-5---clustering.html#cb81-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-8"><a href="capítulo-n-5---clustering.html#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(twogr)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>)</span>
<span id="cb81-9"><a href="capítulo-n-5---clustering.html#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(twogr, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">colour =</span> seq,<span class="at">fill =</span> seq)) <span class="sc">+</span></span>
<span id="cb81-10"><a href="capítulo-n-5---clustering.html#cb81-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hex</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-38-1.png" width="672" /></p>
<p>Simular n=2000variables binarias de longitud len=200que indican la calidad de nsecuenciación de lecturas de longitud len. Para simplificar, supongamos que los errores de secuenciación ocurren de manera independiente y uniforme con probabilidad perr=0.001. Es decir, solo nos importa si una base se llamó correctamente ( TRUE) O no ( FALSE).</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="capítulo-n-5---clustering.html#cb82-1" aria-hidden="true" tabindex="-1"></a>n    <span class="ot">=</span> <span class="dv">2000</span></span>
<span id="cb82-2"><a href="capítulo-n-5---clustering.html#cb82-2" aria-hidden="true" tabindex="-1"></a>len  <span class="ot">=</span> <span class="dv">200</span></span>
<span id="cb82-3"><a href="capítulo-n-5---clustering.html#cb82-3" aria-hidden="true" tabindex="-1"></a>perr <span class="ot">=</span> <span class="fl">0.001</span></span>
<span id="cb82-4"><a href="capítulo-n-5---clustering.html#cb82-4" aria-hidden="true" tabindex="-1"></a>seqs <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">runif</span>(n <span class="sc">*</span> len) <span class="sc">&gt;=</span> perr, <span class="at">nrow =</span> n, <span class="at">ncol =</span> len)</span></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="capítulo-n-5---clustering.html#cb83-1" aria-hidden="true" tabindex="-1"></a>dists <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">dist</span>(seqs, <span class="at">method =</span> <span class="st">&quot;manhattan&quot;</span>))</span></code></pre></div>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="capítulo-n-5---clustering.html#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;tibble&quot;</span>)</span>
<span id="cb84-2"><a href="capítulo-n-5---clustering.html#cb84-2" aria-hidden="true" tabindex="-1"></a>dfseqs <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb84-3"><a href="capítulo-n-5---clustering.html#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">k =</span> <span class="dv">10</span> <span class="sc">^</span> <span class="fu">seq</span>(<span class="fu">log10</span>(<span class="dv">2</span>), <span class="fu">log10</span>(n), <span class="at">length.out =</span> <span class="dv">20</span>),</span>
<span id="cb84-4"><a href="capítulo-n-5---clustering.html#cb84-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">diameter =</span> <span class="fu">vapply</span>(k, <span class="cf">function</span>(i) {</span>
<span id="cb84-5"><a href="capítulo-n-5---clustering.html#cb84-5" aria-hidden="true" tabindex="-1"></a>    s <span class="ot">=</span> <span class="fu">sample</span>(n, i)</span>
<span id="cb84-6"><a href="capítulo-n-5---clustering.html#cb84-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">max</span>(dists[s, s])</span>
<span id="cb84-7"><a href="capítulo-n-5---clustering.html#cb84-7" aria-hidden="true" tabindex="-1"></a>    }, <span class="fu">numeric</span>(<span class="dv">1</span>)))</span>
<span id="cb84-8"><a href="capítulo-n-5---clustering.html#cb84-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dfseqs, <span class="fu">aes</span>(<span class="at">x =</span> k, <span class="at">y =</span> diameter)) <span class="sc">+</span> <span class="fu">geom_point</span>()<span class="sc">+</span><span class="fu">geom_smooth</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-41-1.png" width="672" />
### Inferir variantes de secuencia</p>
<p>Las secuencias desreplicadas se leen y luego se ejecuta la eliminación de ruido divisiva y la estimación con el dadafuncionan como en el siguiente código.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="capítulo-n-5---clustering.html#cb85-1" aria-hidden="true" tabindex="-1"></a>simseq10K <span class="ot">=</span> <span class="fu">replicate</span>(<span class="fl">1e5</span>, <span class="fu">sum</span>(<span class="fu">rpois</span>(<span class="dv">200</span>, <span class="fl">0.0005</span>)))</span>
<span id="cb85-2"><a href="capítulo-n-5---clustering.html#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(simseq10K)</span></code></pre></div>
<pre><code>## [1] 0.10143</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="capítulo-n-5---clustering.html#cb87-1" aria-hidden="true" tabindex="-1"></a>vcd<span class="sc">::</span><span class="fu">distplot</span>(simseq10K, <span class="st">&quot;poisson&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-42-1.png" width="672" /></p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="capítulo-n-5---clustering.html#cb88-1" aria-hidden="true" tabindex="-1"></a>derepFs <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="at">file=</span><span class="st">&quot;../data/derepFs.rds&quot;</span>)</span>
<span id="cb88-2"><a href="capítulo-n-5---clustering.html#cb88-2" aria-hidden="true" tabindex="-1"></a>derepRs <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="at">file=</span><span class="st">&quot;../data/derepRs.rds&quot;</span>)</span>
<span id="cb88-3"><a href="capítulo-n-5---clustering.html#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;dada2&quot;</span>)</span>
<span id="cb88-4"><a href="capítulo-n-5---clustering.html#cb88-4" aria-hidden="true" tabindex="-1"></a>ddF <span class="ot">=</span> <span class="fu">dada</span>(derepFs, <span class="at">err =</span> <span class="cn">NULL</span>, <span class="at">selfConsist =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Initializing error rates to maximum possible estimate.
## selfConsist step 1 ....................
##    selfConsist step 2
##    selfConsist step 3
##    selfConsist step 4
##    selfConsist step 5
##    selfConsist step 6
## Convergence after  6  rounds.</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="capítulo-n-5---clustering.html#cb90-1" aria-hidden="true" tabindex="-1"></a>ddR <span class="ot">=</span> <span class="fu">dada</span>(derepRs, <span class="at">err =</span> <span class="cn">NULL</span>, <span class="at">selfConsist =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Initializing error rates to maximum possible estimate.
## selfConsist step 1 ....................
##    selfConsist step 2
##    selfConsist step 3
##    selfConsist step 4
##    selfConsist step 5
##    selfConsist step 6
##    selfConsist step 7
## Convergence after  7  rounds.</code></pre>
<p>Para verificar que las tasas de transición de error se hayan estimado razonablemente bien, inspeccionamos el ajuste entre las tasas de error observadas (puntos negros) y las tasas de error ajustadas (líneas negras)</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="capítulo-n-5---clustering.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotErrors</span>(ddF)</span></code></pre></div>
<pre><code>## Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
<p><img src="_main_files/figure-html/unnamed-chunk-44-1.png" width="672" /></p>
<p>Una vez que se han estimado los errores, el algoritmo se vuelve a ejecutar en los datos para encontrar las variantes de la secuencia:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="capítulo-n-5---clustering.html#cb94-1" aria-hidden="true" tabindex="-1"></a>dadaFs <span class="ot">=</span> <span class="fu">dada</span>(derepFs, <span class="at">err=</span>ddF[[<span class="dv">1</span>]]<span class="sc">$</span>err_out, <span class="at">pool =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 20 samples were pooled: 139327 reads in 22977 unique sequences.</code></pre>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="capítulo-n-5---clustering.html#cb96-1" aria-hidden="true" tabindex="-1"></a>dadaRs <span class="ot">=</span> <span class="fu">dada</span>(derepRs, <span class="at">err=</span>ddR[[<span class="dv">1</span>]]<span class="sc">$</span>err_out, <span class="at">pool =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 20 samples were pooled: 139327 reads in 19360 unique sequences.</code></pre>
<p>La inferencia de secuencias elimina casi todas las sustituciones y errores de los datos. Fusionamos las secuencias directa e inversa inferidas, mientras eliminamos las secuencias emparejadas que no se superponen perfectamente como control final contra los errores residuales.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="capítulo-n-5---clustering.html#cb98-1" aria-hidden="true" tabindex="-1"></a>mergers <span class="ot">=</span> <span class="fu">mergePairs</span>(dadaFs, derepFs, dadaRs, derepRs)</span></code></pre></div>
<p>Producimos una tabla de contingencia.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="capítulo-n-5---clustering.html#cb99-1" aria-hidden="true" tabindex="-1"></a>seqtab.all <span class="ot">=</span> <span class="fu">makeSequenceTable</span>(mergers[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;Mock&quot;</span>,<span class="fu">names</span>(mergers))])</span></code></pre></div>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="capítulo-n-5---clustering.html#cb100-1" aria-hidden="true" tabindex="-1"></a>dadada <span class="ot">=</span> <span class="fu">unique</span>(<span class="fu">vapply</span>(dadaRs, class, <span class="fu">character</span>(<span class="dv">1</span>)))</span>
<span id="cb100-2"><a href="capítulo-n-5---clustering.html#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">is.list</span>(dadaRs), <span class="fu">identical</span>(<span class="st">&quot;dada&quot;</span>, dadada))</span></code></pre></div>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="capítulo-n-5---clustering.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(dadaRs)</span></code></pre></div>
<pre><code>## [1] 20</code></pre>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="capítulo-n-5---clustering.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(dadaFs)</span></code></pre></div>
<pre><code>## [1] 20</code></pre>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="capítulo-n-5---clustering.html#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dadaRs)</span></code></pre></div>
<pre><code>## [1] &quot;list&quot;</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="capítulo-n-5---clustering.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dadaRs)</span></code></pre></div>
<pre><code>##  [1] &quot;F3D0&quot;   &quot;F3D1&quot;   &quot;F3D141&quot; &quot;F3D142&quot; &quot;F3D143&quot; &quot;F3D144&quot; &quot;F3D145&quot; &quot;F3D146&quot;
##  [9] &quot;F3D147&quot; &quot;F3D148&quot; &quot;F3D149&quot; &quot;F3D150&quot; &quot;F3D2&quot;   &quot;F3D3&quot;   &quot;F3D5&quot;   &quot;F3D6&quot;  
## [17] &quot;F3D7&quot;   &quot;F3D8&quot;   &quot;F3D9&quot;   &quot;Mock&quot;</code></pre>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="capítulo-n-5---clustering.html#cb109-1" aria-hidden="true" tabindex="-1"></a>mergers <span class="ot">=</span> <span class="fu">mergePairs</span>(dadaFs, derepFs, dadaRs, derepRs)</span>
<span id="cb109-2"><a href="capítulo-n-5---clustering.html#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mergers)</span></code></pre></div>
<pre><code>## [1] &quot;list&quot;</code></pre>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="capítulo-n-5---clustering.html#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(mergers)</span></code></pre></div>
<pre><code>## [1] 20</code></pre>
<p><strong>Pregunta</strong></p>
<p>Explora los componentes de los objetos. dadaRsy mergers.</p>
<p>as quimeras son secuencias que se crean artificialmente durante la amplificación por PCR mediante la fusión de dos (en casos raros, más) de las secuencias originales. Para completar nuestro flujo de trabajo de eliminación de ruido, los eliminamos con una llamada a la función removeBimeraDenovo, dejándonos con una tabla de contingencia limpia que usaremos más adelante.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="capítulo-n-5---clustering.html#cb113-1" aria-hidden="true" tabindex="-1"></a>seqtab <span class="ot">=</span> <span class="fu">removeBimeraDenovo</span>(seqtab.all)</span></code></pre></div>
</div>
</div>
<div id="ejercicios-2" class="section level2 hasAnchor" number="5.9">
<h2><span class="header-section-number">5.9</span> Ejercicios<a href="capítulo-n-5---clustering.html#ejercicios-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Calcular el índice de silueta para el simdatdatos que simulamos.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="capítulo-n-5---clustering.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;cluster&quot;</span>)</span>
<span id="cb114-2"><a href="capítulo-n-5---clustering.html#cb114-2" aria-hidden="true" tabindex="-1"></a>pam4 <span class="ot">=</span> <span class="fu">pam</span>(simdatxy, <span class="dv">4</span>)</span>
<span id="cb114-3"><a href="capítulo-n-5---clustering.html#cb114-3" aria-hidden="true" tabindex="-1"></a>sil <span class="ot">=</span> <span class="fu">silhouette</span>(pam4, <span class="dv">4</span>)</span>
<span id="cb114-4"><a href="capítulo-n-5---clustering.html#cb114-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sil, <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;purple&quot;</span>), <span class="at">main=</span><span class="st">&quot;Silhouette&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-51-1.png" width="672" /></p>
<p>Hacer una representación de “caracteres” de la distancia entre las 20 ubicaciones en el dunedatos del vegano paquete symnum.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="capítulo-n-5---clustering.html#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(<span class="st">&quot;kernlab&quot;</span>)</span>
<span id="cb115-2"><a href="capítulo-n-5---clustering.html#cb115-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;spirals&quot;</span>)</span>
<span id="cb115-3"><a href="capítulo-n-5---clustering.html#cb115-3" aria-hidden="true" tabindex="-1"></a>clusts <span class="ot">=</span> <span class="fu">kmeans</span>(spirals,<span class="dv">2</span>)<span class="sc">$</span>cluster</span>
<span id="cb115-4"><a href="capítulo-n-5---clustering.html#cb115-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spirals, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>)[clusts])</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-52-1.png" width="672" /></p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="capítulo-n-5---clustering.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;spirals&quot;</span>, <span class="at">package =</span> <span class="st">&quot;kernlab&quot;</span>)</span>
<span id="cb116-2"><a href="capítulo-n-5---clustering.html#cb116-2" aria-hidden="true" tabindex="-1"></a>res.dbscan <span class="ot">=</span> dbscan<span class="sc">::</span><span class="fu">dbscan</span>(spirals, <span class="at">eps =</span> <span class="fl">0.16</span>, <span class="at">minPts =</span> <span class="dv">3</span>)</span>
<span id="cb116-3"><a href="capítulo-n-5---clustering.html#cb116-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spirals,<span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;forestgreen&quot;</span>)[res.dbscan<span class="sc">$</span>cluster])</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-52-2.png" width="672" /></p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="capítulo-n-5---clustering.html#cb117-1" aria-hidden="true" tabindex="-1"></a>sc <span class="ot">=</span> <span class="fu">specc</span>(spirals, <span class="at">centers=</span><span class="dv">2</span>)</span>
<span id="cb117-2"><a href="capítulo-n-5---clustering.html#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spirals, <span class="at">col=</span>sc)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-53-1.png" width="672" /></p>
<p>Bioinformática de amplicón: desde lecturas sin procesar hasta secuencias desreplicadas</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="capítulo-n-5---clustering.html#cb118-1" aria-hidden="true" tabindex="-1"></a>base_dir <span class="ot">=</span> <span class="st">&quot;../data&quot;</span></span>
<span id="cb118-2"><a href="capítulo-n-5---clustering.html#cb118-2" aria-hidden="true" tabindex="-1"></a>miseq_path <span class="ot">=</span> <span class="fu">file.path</span>(base_dir, <span class="st">&quot;MiSeq_SOP&quot;</span>)</span>
<span id="cb118-3"><a href="capítulo-n-5---clustering.html#cb118-3" aria-hidden="true" tabindex="-1"></a>filt_path <span class="ot">=</span> <span class="fu">file.path</span>(miseq_path, <span class="st">&quot;filtered&quot;</span>)</span>
<span id="cb118-4"><a href="capítulo-n-5---clustering.html#cb118-4" aria-hidden="true" tabindex="-1"></a>fnFs <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(miseq_path, <span class="at">pattern=</span><span class="st">&quot;_R1_001.fastq&quot;</span>))</span>
<span id="cb118-5"><a href="capítulo-n-5---clustering.html#cb118-5" aria-hidden="true" tabindex="-1"></a>fnRs <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(miseq_path, <span class="at">pattern=</span><span class="st">&quot;_R2_001.fastq&quot;</span>))</span>
<span id="cb118-6"><a href="capítulo-n-5---clustering.html#cb118-6" aria-hidden="true" tabindex="-1"></a>sampleNames <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(fnFs, <span class="st">&quot;_&quot;</span>), <span class="st">`</span><span class="at">[</span><span class="st">`</span>, <span class="dv">1</span>)</span>
<span id="cb118-7"><a href="capítulo-n-5---clustering.html#cb118-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file_test</span>(<span class="st">&quot;-d&quot;</span>, filt_path)) <span class="fu">dir.create</span>(filt_path)</span>
<span id="cb118-8"><a href="capítulo-n-5---clustering.html#cb118-8" aria-hidden="true" tabindex="-1"></a>filtFs <span class="ot">=</span> <span class="fu">file.path</span>(filt_path, <span class="fu">paste0</span>(sampleNames, <span class="st">&quot;_F_filt.fastq.gz&quot;</span>))</span>
<span id="cb118-9"><a href="capítulo-n-5---clustering.html#cb118-9" aria-hidden="true" tabindex="-1"></a>filtRs <span class="ot">=</span> <span class="fu">file.path</span>(filt_path, <span class="fu">paste0</span>(sampleNames, <span class="st">&quot;_R_filt.fastq.gz&quot;</span>))</span>
<span id="cb118-10"><a href="capítulo-n-5---clustering.html#cb118-10" aria-hidden="true" tabindex="-1"></a>fnFs <span class="ot">=</span> <span class="fu">file.path</span>(miseq_path, fnFs)</span>
<span id="cb118-11"><a href="capítulo-n-5---clustering.html#cb118-11" aria-hidden="true" tabindex="-1"></a>fnRs <span class="ot">=</span> <span class="fu">file.path</span>(miseq_path, fnRs)</span>
<span id="cb118-12"><a href="capítulo-n-5---clustering.html#cb118-12" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">length</span>(fnFs))</span></code></pre></div>
<pre><code>## [1] 20</code></pre>
<p>Tendremos que filtrar las lecturas de baja calidad y recortarlas a una longitud constante.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="capítulo-n-5---clustering.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotQualityProfile</span>(fnFs[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Forward&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-55-1.png" width="672" /></p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="capítulo-n-5---clustering.html#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotQualityProfile</span>(fnRs[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Reverse&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-55-2.png" width="672" /></p>
<p>Genere gráficos similares para cuatro conjuntos seleccionados al azar de lecturas directas e inversas. Compare las cualidades de lectura directa e inversa; ¿que notaste?</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="capítulo-n-5---clustering.html#cb122-1" aria-hidden="true" tabindex="-1"></a>ii <span class="ot">=</span> <span class="fu">sample</span>(<span class="fu">length</span>(fnFs), <span class="dv">4</span>)</span>
<span id="cb122-2"><a href="capítulo-n-5---clustering.html#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotQualityProfile</span>(fnFs[ii]) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Forward&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-56-1.png" width="672" /></p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="capítulo-n-5---clustering.html#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotQualityProfile</span>(fnRs[ii]) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Reverse&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-56-2.png" width="672" /></p>
<p>Aquí, las lecturas directas mantienen una alta calidad en todo momento, mientras que la calidad de las lecturas inversas cae significativamente alrededor de la posición 160. Por lo tanto, truncamos las lecturas directas en la posición 240 y recortamos los primeros 10 nucleótidos ya que estas posiciones son de menor calidad. Las lecturas inversas se recortan en la posición 160. Combine estos parámetros de recorte con los parámetros de filtrado estándar y recuerde imponer un máximo de 2 errores esperados por lectura.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="capítulo-n-5---clustering.html#cb124-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">=</span> <span class="fu">filterAndTrim</span>(fnFs, filtFs, fnRs, filtRs, <span class="at">truncLen=</span><span class="fu">c</span>(<span class="dv">240</span>,<span class="dv">160</span>),</span>
<span id="cb124-2"><a href="capítulo-n-5---clustering.html#cb124-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">maxN=</span><span class="dv">0</span>, <span class="at">maxEE=</span><span class="dv">2</span>, <span class="at">truncQ=</span><span class="dv">2</span>, <span class="at">rm.phix=</span><span class="cn">TRUE</span>,  <span class="at">trimLeft=</span><span class="dv">10</span>,</span>
<span id="cb124-3"><a href="capítulo-n-5---clustering.html#cb124-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">compress=</span><span class="cn">TRUE</span>, <span class="at">multithread=</span><span class="cn">TRUE</span>) <span class="co"># On Windows set multithread=FALSE</span></span>
<span id="cb124-4"><a href="capítulo-n-5---clustering.html#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(out)</span></code></pre></div>
<pre><code>##                               reads.in reads.out
## F3D0_S188_L001_R1_001.fastq       7793      7139
## F3D1_S189_L001_R1_001.fastq       5869      5314
## F3D141_S207_L001_R1_001.fastq     5958      5478
## F3D142_S208_L001_R1_001.fastq     3183      2926
## F3D143_S209_L001_R1_001.fastq     3178      2955
## F3D144_S210_L001_R1_001.fastq     4827      4323</code></pre>
<p>Los datos de la secuencia se importaron a R desde fastq archivos fastq para cada muestra) y se desreplicaron simultáneamente para eliminar la redundancia.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="capítulo-n-5---clustering.html#cb126-1" aria-hidden="true" tabindex="-1"></a>derepFs <span class="ot">=</span> <span class="fu">derepFastq</span>(filtFs, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb126-2"><a href="capítulo-n-5---clustering.html#cb126-2" aria-hidden="true" tabindex="-1"></a>derepRs <span class="ot">=</span> <span class="fu">derepFastq</span>(filtRs, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb126-3"><a href="capítulo-n-5---clustering.html#cb126-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(derepFs) <span class="ot">=</span> sampleNames</span>
<span id="cb126-4"><a href="capítulo-n-5---clustering.html#cb126-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(derepRs) <span class="ot">=</span> sampleNames</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="capítulo-n-4---modelos-de-mezcla.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="footnotes-and-citations.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/04-C5-Clustering.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
